---
title: "B. Using R to Understand Bioinformatic Results"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{B. Using R to Understand Bioinformatic Results}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This workshop introduces *R* as an essential tool in exploratory
analysis of bioinformatic data. No previous experience with *R* is
required. We start with a very short introduction to *R*, mentioning
vectors, functions, and the `data.frame` for representing tabular
data. We explore some essential data management tasks, for instance
summarizing cell types and plotting a 'UMAP' in a single-cell RNASeq
experiment. We adopt the 'tidy' paradigm, using the [dplyr][] package
for data management and [ggplot2][] for data visualization. The
workshop concludes with a short tour of approaches to enhance
reproducible research in our day-to-day work.

[dplyr]: https://cran.r-project.org/package=dplyr
[ggplot2]: https://cran.r-project.org/package=ggplot2

# Introduction to *R*

Vectors

- Common types of vectors include character, numeric, logical, ... Can
  include `NA` to indicate that the information is not available.

    ```{r vectors}
    name <- c("Benilton", "Martin")
    height_cm <- c(NA, 185)
    is_brazillian <- c(TRUE, FALSE)
    ```

- Vectorized logical and subsetting

  - Using vectors can be very powerful, for instance to find the name
    of Brazillians does not require a `for` loop or other iteration.
    
    ```{r name}
    name[is_brazillian]
    ```

  - According to [Google][], the average waist size in men is
    102cm. Approximating the human shape as a cylinder, then the
    volume of Benilton and Martin is

    ```{r volume}
    waist_diameter_cm <- 102
    ## d = 2 pi r, so r = d / (2 pi)
    radius_cm <- waist_diameter_cm / (2 * pi)
    ## V = pi r^2 h
    volume_cm3 <- pi * radius_cm^2 * height_cm
    volume_cm3
    ```

  - We didn't know Benilton's height, so we can't know his
    volume. Apparently, by these calculations, Martin has an expected
    volume in excess of 150 thousand cubic centimeters.

[Google]: https://www.google.com/search?q=average+waist+size+for+men+in+cm

Functions

`data.frame()`

- Key concept for organizing variables (columns) observed on samples
  (rows).

    ```{r data.frame}
    df <- data.frame(
        name = name,
        height = height_cm,
        volume = volume_cm3,
        is_brazillian = is_brazillian
    )
    df
    ```

- Typically, `data.frame` is the result of reading data from a file,
  e.g., `read.csv()`.

# 'tidy' analysis

This section makes use of [dplyr][], [readr][], [ggplot2][], and
[plotly][]; make sure these are installed

[readr]: https://cran.r-project.org/package=readr
[plotly]: https://cran.r-project.org/package=plotly

```{r tidy-pkgs, message = FALSE}
pkgs <- c("dplyr", "readr", "ggplot2", "plotly")
need <- pkgs[!pkgs %in% rownames(installed.packages())]
if (length(need))
    install.packages(need, repos = "https://cran.r-project.org")
```

## Tidy analysis and visualization

Why do I like 'tidy' analysis?

- A single representation -- the 'tibble' (aka 'data.frame') -- covers
  many typical data analysis tasks, and provides a format to shoot for
  when tempted by more complicated data structures
- A few 'verbs' with consistent signatures help circumscribe the data
  transformations available.
- The 'pipe' operator `|>` helps structure transformations into an
  easy-to-understand order.

```{r tidy-example}
library(dplyr)

## a familiar built-in dataset, represented as a `tibble()`...
mtcars_tbl <-
    mtcars |>
    as_tibble(rownames = "model")
mtcars_tbl
```

The [dplyr][] package is at the center of 'tidy' analysis

- `tibble()` -- representation of a `data.frame`, with better display
  of long and wide data frames. `tribble()` constructs a tibble in a
  way that makes the relationship between data across rows more
  transparent.
- `glimpse()` -- providing a quick look into the columns and data in
  the tibble by transposing the tibble and display each 'column' on a
  single line.
- `select()` -- column selection.
- `filter()`, `slice()` -- row selection.
- `pull()` -- extract a single column as a vector.
- `mutate()` -- column transformation.
- `count()` -- count occurences in one or more columns.
- `arrange()` -- order rows by values in one or more columns.
- `distinct()` -- reduce a tibble to only unique rows.
- `group_by()` -- perform computations on groups defined by one or
  several columns.
- `summarize()` -- calculate summary statstics for groups.
- `left_join()`, `right_join()`, `inner_join()` -- merge two tibbles
  based on shared columns, preserving all rows in the first
  (`left_join()`) or second (`right_join()`) or both (`inner_join()`)
  tibble.

```{r}
## count # of cars with 4, 6, ... cylinders
mtcars_tbl |>
    count(cyl)

## select specific columns and filter rows based on 'cyl' and 'mpg'
mtcars_tbl |>
    select(model, mpg, cyl, disp) |>
    filter(
        cyl >= 8,
        mpg > 15
    )
```

Visualization

- A key to communicating results
- [ggplot2][] provides a 'grammar of graphics' to produce high-quality
  graphics.

```{r}
library(ggplot2)
plt <-
    ggplot(mtcars_tbl) +
    aes(x = factor(cyl), y = mpg) +
    geom_boxplot() +
    geom_jitter()
plt
```

- [plotly][] provides interactivity (e.g., 'mouse over' a point to see
  the details of the point), which can be very useful when trying to
  communicate results in an exploratory manner.

```{r, message = FALSE}
library(plotly)
plt <-
    ggplot(mtcars_tbl) +
    aes(x = factor(cyl), y = mpg, text = model) +
    geom_jitter(width = .25) +
    geom_boxplot()
ggplotly(plt)
```

# Case study: single-cell UMAP

Load a 'csv' (comma-separated value) file prepared with the a 'UMAP'.

```{r, eval = FALSE}
## use `file.choose()` or similar for your own data sets
umap_csv_file <- system.file(package= "XM2023", "extdata", "umap.csv")
umap_data <- readr::read_csv(umap_csv_file)
umap_data
```

# Reproducibility

Packages

- Provide a robust way to document requirements for analysis, and to
  organize complicated analyses into distinct steps.

    ```{r devtools, eval = FALSE}
    devtools::create("MyAnalysis")
    setwd("MyAnalysis")
    usethis::use_vignette("a_data_management", "A. Data management")
    usethis::use_vignette("b_exploratory_visualization", "B. Exploration")
    ```

- I've found [pkgdown][] to be very useful for presenting packages to
  my users. For instance, [material from this workshop][workshop] is
  available through pkgdown.

    ```{r pkgdown, eval = FALSE}
    usethis::use_pkgdown()
    pkgdown::build_site()
    ```

[pkgdown]: https://cran.r-project.org/package=pkgdown
[workshop]: https://mtmorgan.github.io/XM2023MM

Vignettes

- Explict description of analysis steps (good for the
  bioinformatician), coupled with text and graphics (good for the
  collaboration).

Git

- Incremental 'commits' as analysis progresses.
- Commits allow confident exploration -- the last commit is always
  available to 'start over'.
- Tags allow checkpointing an analysis, e.g., the version of the
  analysis used in the original manuscript submission; the version of
  the analysis associated with the revision and final publciation.

Containers

- A fully reproducible analysis is very challenging to implement --
  specifying software version is not enough, and is not easy for a
  future investigator to re-establish.
- Containers like [docker][] or [singularity][] provide one mechanism
  for creating a 'snapshot' capturing exactly the software used.
- Beware! Complicated containers might result in a fully reproducible
  analysis, but provide little confidence in the robustness of the
  analysis.

[docker]: https://www.docker.com
[singularity]: https://sylabs.io/

# Session information

This document was produced with the following *R* software:

```{r session_info}
sessionInfo()
```
