---
title: "C.0 R and Bioconductor for Genomic Analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{C.0 R and Bioconductor for Genomic Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This workshop walks through a single-cellsis with Bioconductor][OSCA]
(OSCA). The workshop assumes some familiarity with *R*, and sufficient
domain knowledge to know at a superficial level the technical details
and scientific motivation for single-cell analysis. The main goal is
to increase participants' confidence in using *R* to embark on
creative and critical data exploration -- analysis of single cell data
is seldom straight-forward, requiring understanding of methods and
critical assessment of data at many steps.

There are three parts to the workshop, probably we will not have time
to work through all. The first provides a high-level overview of how
data available in the CELLxGENE portal can be used in the [Seurat][]
(Seurat is not a _Bioconductor_ package) and [SingleCellExperiment][]
/ _Bioconductor_ ecosystem.

- [CELLxGENE, _Seurat_ and _Bioconductor_][C-1]

The remaining parts walk through two chapters of [OSCA][], chosen as
much because I wanted to learn a bit more about these steps as for any
other reason.

- [Annotating cell types][C-2]
- [Differential expression][C-3]

[OSCA]: https://bioconductor.org/books/release/OSCA/
[Seurat]: https://satijalab.org/seurat/index.html
[SingleCellExperiment]: https://bioconductor.org/packages/SingleCellExperiment


## Differential expression

This script is derived from the [OSCA Multi-Sample][OSCA-M] book,
[Chapter 4][OSCA-M-4]: DE analyses between conditions. See the book
for full details.

```{r, echo = FALSE, message = FALSE}
library(MouseGastrulationData)
library(scater)
library(scran)
library(batchelor)
requireNamespace("igraph")
```

Initial analysis

```{r, message = FALSE}
## load data
library(MouseGastrulationData)
sce.chimera <- WTChimeraData(samples=5:10)
sce.chimera

## feature annotation
library(scater)
rownames(sce.chimera) <- uniquifyFeatureNames(
    rowData(sce.chimera)$ENSEMBL, rowData(sce.chimera)$SYMBOL
)

## qc
drop <- sce.chimera$celltype.mapped %in% c("stripped", "Doublet")
sce.chimera <- sce.chimera[,!drop]

## normalization
sce.chimera <- logNormCounts(sce.chimera)

## variance modeling
library(scran)
dec.chimera <- modelGeneVar(sce.chimera, block=sce.chimera$sample)
chosen.hvgs <- dec.chimera$bio > 0

## merge batches
library(batchelor)
set.seed(01001001)
merged <- correctExperiments(
    sce.chimera, 
    batch=sce.chimera$sample, 
    subset.row=chosen.hvgs,
    PARAM=FastMnnParam(
        merge.order=list(
            list(1,3,5), # WT (3 replicates)
            list(2,4,6)  # td-Tomato (3 replicates)
        )
    )
)

## cluster
g <- buildSNNGraph(merged, use.dimred="corrected")
clusters <- igraph::cluster_louvain(g)
colLabels(merged) <- factor(clusters$membership)

## dimensionality reduction
merged <- runTSNE(merged, dimred="corrected", external_neighbors=TRUE)
merged <- runUMAP(merged, dimred="corrected", external_neighbors=TRUE)
```

[OSCA-M]: https://bioconductor.org/books/OSCA.multisample/
[OSCA-M-4]: https://bioconductor.org/books/release/OSCA.multisample/multi-sample-comparisons.html

# Session information

This document was produced with the following *R* software:

```{r session_info}
sessionInfo()
```

