<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>B. Using R to Understand Bioinformatic Results • XM2023</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="B. Using R to Understand Bioinformatic Results">
<meta property="og:description" content="XM2023">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">XM2023</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9001</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a_talk.html">A. How Bioconductor Advances Science</a>
    </li>
    <li>
      <a href="../articles/b_course_part_1.html">B. Using R to Understand Bioinformatic Results</a>
    </li>
    <li>
      <a href="../articles/c_course_part_2.html">C. R / Bioconductor for Genomic Analysis</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/mtmorgan/XM2023" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>B. Using R to Understand Bioinformatic
Results</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mtmorgan/XM2023MM/blob/HEAD/vignettes/b_course_part_1.Rmd" class="external-link"><code>vignettes/b_course_part_1.Rmd</code></a></small>
      <div class="hidden name"><code>b_course_part_1.Rmd</code></div>

    </div>

    
    
<p>This workshop introduces <em>R</em> as an essential tool in
exploratory analysis of bioinformatic data. No previous experience with
<em>R</em> is required. We start with a very short introduction to
<em>R</em>, mentioning vectors, functions, and the
<code>data.frame</code> for representing tabular data. We explore some
essential data management tasks, for instance summarizing cell types and
plotting a ‘UMAP’ in a single-cell RNASeq experiment. We adopt the
‘tidy’ paradigm, using the <a href="https://cran.r-project.org/package=dplyr" class="external-link">dplyr</a> package for
data management and <a href="https://cran.r-project.org/package=ggplot2" class="external-link">ggplot2</a> for data
visualization. The workshop concludes with a short tour of approaches to
enhance reproducible research in our day-to-day work.</p>
<div class="section level2">
<h2 id="introduction-to-r">Introduction to <em>R</em><a class="anchor" aria-label="anchor" href="#introduction-to-r"></a>
</h2>
<p>Vectors</p>
<ul>
<li>
<p>Common types of vectors include character, numeric, logical, …
Can include <code>NA</code> to indicate that the information is not
available.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Benilton"</span>, <span class="st">"Martin"</span><span class="op">)</span></span>
<span><span class="va">height_cm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fl">185</span><span class="op">)</span></span>
<span><span class="va">is_brazillian</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p>Vectorized logical and subsetting</p>
<ul>
<li>
<p>Using vectors can be very powerful, for instance to find the name
of Brazillians does not require a <code>for</code> loop or other
iteration.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">name</span><span class="op">[</span><span class="va">is_brazillian</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] "Benilton"</span></span></code></pre></div>
</li>
<li>
<p>According to <a href="https://www.google.com/search?q=average+waist+size+for+men+in+cm" class="external-link">Google</a>,
the average waist size in men is 102cm. Approximating the human shape as
a cylinder, then the volume of Benilton and Martin is</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">waist_diameter_cm</span> <span class="op">&lt;-</span> <span class="fl">102</span></span>
<span><span class="co">## d = 2 pi r, so r = d / (2 pi)</span></span>
<span><span class="va">radius_cm</span> <span class="op">&lt;-</span> <span class="va">waist_diameter_cm</span> <span class="op">/</span> <span class="op">(</span><span class="fl">2</span> <span class="op">*</span> <span class="va">pi</span><span class="op">)</span></span>
<span><span class="co">## V = pi r^2 h</span></span>
<span><span class="va">volume_cm3</span> <span class="op">&lt;-</span> <span class="va">pi</span> <span class="op">*</span> <span class="va">radius_cm</span><span class="op">^</span><span class="fl">2</span> <span class="op">*</span> <span class="va">height_cm</span></span>
<span><span class="va">volume_cm3</span></span>
<span><span class="co">#&gt; [1]       NA 153165.9</span></span></code></pre></div>
</li>
<li><p>We didn’t know Benilton’s height, so we can’t know his volume.
Apparently, by these calculations, Martin has an expected volume in
excess of 150 thousand cubic centimeters.</p></li>
</ul>
</li>
</ul>
<p>Functions</p>
<p><code><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame()</a></code></p>
<ul>
<li>
<p>Key concept for organizing variables (columns) observed on
samples (rows).</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>    name <span class="op">=</span> <span class="va">name</span>,</span>
<span>    height <span class="op">=</span> <span class="va">height_cm</span>,</span>
<span>    volume <span class="op">=</span> <span class="va">volume_cm3</span>,</span>
<span>    is_brazillian <span class="op">=</span> <span class="va">is_brazillian</span></span>
<span><span class="op">)</span></span>
<span><span class="va">df</span></span>
<span><span class="co">#&gt;       name height   volume is_brazillian</span></span>
<span><span class="co">#&gt; 1 Benilton     NA       NA          TRUE</span></span>
<span><span class="co">#&gt; 2   Martin    185 153165.9         FALSE</span></span></code></pre></div>
</li>
<li><p>Typically, <code>data.frame</code> is the result of reading data
from a file, e.g., <code><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv()</a></code>.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="tidy-analysis">‘tidy’ analysis<a class="anchor" aria-label="anchor" href="#tidy-analysis"></a>
</h2>
<p>This section makes use of <a href="https://cran.r-project.org/package=dplyr" class="external-link">dplyr</a>, <a href="https://cran.r-project.org/package=readr" class="external-link">readr</a>, <a href="https://cran.r-project.org/package=ggplot2" class="external-link">ggplot2</a>, and <a href="https://cran.r-project.org/package=plotly" class="external-link">plotly</a>; make sure
these are installed</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pkgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dplyr"</span>, <span class="st">"readr"</span>, <span class="st">"ggplot2"</span>, <span class="st">"plotly"</span><span class="op">)</span></span>
<span><span class="va">need</span> <span class="op">&lt;-</span> <span class="va">pkgs</span><span class="op">[</span><span class="op">!</span><span class="va">pkgs</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">need</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="va">need</span>, repos <span class="op">=</span> <span class="st">"https://cran.r-project.org"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="tidy-analysis-and-visualization">Tidy analysis and visualization<a class="anchor" aria-label="anchor" href="#tidy-analysis-and-visualization"></a>
</h3>
<p>Why do I like ‘tidy’ analysis?</p>
<ul>
<li>A single representation – the ‘tibble’ (aka ‘data.frame’) – covers
many typical data analysis tasks, and provides a format to shoot for
when tempted by more complicated data structures</li>
<li>A few ‘verbs’ with consistent signatures help circumscribe the data
transformations available.</li>
<li>The ‘pipe’ operator <code>|&gt;</code> helps structure
transformations into an easy-to-understand order.</li>
</ul>
<p>The <a href="https://cran.r-project.org/package=dplyr" class="external-link">dplyr</a>
package is at the center of ‘tidy’ analysis</p>
<ul>
<li>
<code>tibble()</code> – representation of a <code>data.frame</code>,
with better display of long and wide data frames. <code>tribble()</code>
constructs a tibble in a way that makes the relationship between data
across rows more transparent.</li>
<li>
<code>glimpse()</code> – providing a quick look into the columns and
data in the tibble by transposing the tibble and display each ‘column’
on a single line.</li>
<li>
<code>select()</code> – column selection.</li>
<li>
<code><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter()</a></code>, <code>slice()</code> – row selection.</li>
<li>
<code>pull()</code> – extract a single column as a vector.</li>
<li>
<code>mutate()</code> – column transformation.</li>
<li>
<code>count()</code> – count occurences in one or more columns.</li>
<li>
<code>arrange()</code> – order rows by values in one or more
columns.</li>
<li>
<code>distinct()</code> – reduce a tibble to only unique rows.</li>
<li>
<code>group_by()</code> – perform computations on groups defined by
one or several columns.</li>
<li>
<code>summarize()</code> – calculate summary statstics for
groups.</li>
<li>
<code>left_join()</code>, <code>right_join()</code>,
<code>inner_join()</code> – merge two tibbles based on shared columns,
preserving all rows in the first (<code>left_join()</code>) or second
(<code>right_join()</code>) or both (<code>inner_join()</code>)
tibble.</li>
</ul>
<p>Visualization</p>
<ul>
<li>A key to communicating results</li>
<li>
<a href="https://cran.r-project.org/package=ggplot2" class="external-link">ggplot2</a>
provides a ‘grammar of graphics’ to produce high-quality graphics.</li>
<li>
<a href="https://cran.r-project.org/package=plotly" class="external-link">plotly</a>
provides interactivity (e.g., ‘mouse over’ a point to see the details of
the point), which can be very useful when trying to communicate results
in an exploratory manner.</li>
</ul>
</div>
<div class="section level3">
<h3 id="case-study-single-cell-umap">Case study: single-cell UMAP<a class="anchor" aria-label="anchor" href="#case-study-single-cell-umap"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h2>
<p>Packages</p>
<ul>
<li>
<p>Provide a robust way to document requirements for analysis, and
to organize complicated analyses into distinct steps.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/create.html" class="external-link">create</a></span><span class="op">(</span><span class="st">"MyAnalysis"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"MyAnalysis"</span><span class="op">)</span></span>
<span><span class="fu">usethis</span><span class="fu">::</span><span class="fu"><a href="https://usethis.r-lib.org/reference/use_vignette.html" class="external-link">use_vignette</a></span><span class="op">(</span><span class="st">"a_data_management"</span>, <span class="st">"A. Data management"</span><span class="op">)</span></span>
<span><span class="fu">usethis</span><span class="fu">::</span><span class="fu"><a href="https://usethis.r-lib.org/reference/use_vignette.html" class="external-link">use_vignette</a></span><span class="op">(</span><span class="st">"b_exploratory_visualization"</span>, <span class="st">"B. Exploration"</span><span class="op">)</span></span></code></pre></div>
</li>
<li>
<p>I’ve found <a href="https://cran.r-project.org/package=pkgdown" class="external-link">pkgdown</a> to be very
useful for presenting packages to my users. For instance, <a href="https://mtmorgan.github.io/XM2023MM">material from this
workshop</a> is available through pkgdown.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">usethis</span><span class="fu">::</span><span class="fu"><a href="https://usethis.r-lib.org/reference/use_pkgdown.html" class="external-link">use_pkgdown</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu">pkgdown</span><span class="fu">::</span><span class="fu"><a href="https://pkgdown.r-lib.org/reference/build_site.html" class="external-link">build_site</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</li>
</ul>
<p>Vignettes</p>
<ul>
<li>Explict description of analysis steps (good for the
bioinformatician), coupled with text and graphics (good for the
collaboration).</li>
</ul>
<p>Git</p>
<ul>
<li>Incremental ‘commits’ as analysis progresses.</li>
<li>Commits allow confident exploration – the last commit is always
available to ‘start over’.</li>
<li>Tags allow checkpointing an analysis, e.g., the version of the
analysis used in the original manuscript submission; the version of the
analysis associated with the revision and final publciation.</li>
</ul>
<p>Containers</p>
<ul>
<li>A fully reproducible analysis is very challenging to implement –
specifying software version is not enough, and is not easy for a future
investigator to re-establish.</li>
<li>Containers like <a href="https://www.docker.com" class="external-link">docker</a> or <a href="https://sylabs.io/" class="external-link">singularity</a> provide one mechanism for
creating a ‘snapshot’ capturing exactly the software used.</li>
<li>Beware! Complicated containers might result in a fully reproducible
analysis, but provide little confidence in the robustness of the
analysis.</li>
</ul>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>This document was produced with the following <em>R</em>
software:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] vctrs_0.6.2       cli_3.6.1         knitr_1.43        rlang_1.1.1      </span></span>
<span><span class="co">#&gt;  [5] xfun_0.39         stringi_1.7.12    purrr_1.0.1       textshaping_0.3.6</span></span>
<span><span class="co">#&gt;  [9] jsonlite_1.8.4    glue_1.6.2        rprojroot_2.0.3   htmltools_0.5.5  </span></span>
<span><span class="co">#&gt; [13] ragg_1.2.5        sass_0.4.6        rmarkdown_2.22    evaluate_0.21    </span></span>
<span><span class="co">#&gt; [17] jquerylib_0.1.4   fastmap_1.1.1     yaml_2.3.7        lifecycle_1.0.3  </span></span>
<span><span class="co">#&gt; [21] memoise_2.0.1     stringr_1.5.0     compiler_4.3.0    fs_1.6.2         </span></span>
<span><span class="co">#&gt; [25] systemfonts_1.0.4 digest_0.6.31     R6_2.5.1          magrittr_2.0.3   </span></span>
<span><span class="co">#&gt; [29] bslib_0.4.2       tools_4.3.0       pkgdown_2.0.7     cachem_1.0.8     </span></span>
<span><span class="co">#&gt; [33] desc_1.4.2</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Martin Morgan, NIH NHGRI U24HG004059.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
