<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="XM2023">
<title>C.0 R and Bioconductor for Genomic Analysis • XM2023</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="C.0 R and Bioconductor for Genomic Analysis">
<meta property="og:description" content="XM2023">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">XM2023</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9009</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/a_talk.html">A. How Bioconductor Advances Science</a>
    <a class="dropdown-item" href="../articles/b_course_part_1.html">B. Using R to Understand Bioinformatic Results</a>
    <a class="dropdown-item" href="../articles/c_0_course_part_2.html">C.0 R and Bioconductor for Genomic Analysis</a>
    <a class="dropdown-item" href="../articles/c_1_cxg.html">C.1 CELLxGENE, Seurat, and Bioconductor</a>
    <a class="dropdown-item" href="../articles/c_2_osca_cell_annotation.html">C.2 Annotating cell types</a>
    <a class="dropdown-item" href="../articles/c_3_osca_differential_expression.html">C.3 Differential Expression</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mtmorgan/XM2023">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>C.0 R and Bioconductor for Genomic Analysis</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mtmorgan/XM2023/blob/HEAD/vignettes/c_0_course_part_2.Rmd" class="external-link"><code>vignettes/c_0_course_part_2.Rmd</code></a></small>
      <div class="d-none name"><code>c_0_course_part_2.Rmd</code></div>
    </div>

    
    
<p>This workshop walks through a single-cellsis with Bioconductor]<a href="https://bioconductor.org/books/release/OSCA/" class="external-link">OSCA</a> (OSCA). The
workshop assumes some familiarity with <em>R</em>, and sufficient domain
knowledge to know at a superficial level the technical details and
scientific motivation for single-cell analysis. The main goal is to
increase participants’ confidence in using <em>R</em> to embark on
creative and critical data exploration – analysis of single cell data is
seldom straight-forward, requiring understanding of methods and critical
assessment of data at many steps.</p>
<p>There are three parts to the workshop, probably we will not have time
to work through all. The first provides a high-level overview of how
data available in the CELLxGENE portal can be used in the <a href="https://satijalab.org/seurat/index.html" class="external-link">Seurat</a> (Seurat is not
a <em>Bioconductor</em> package) and <a href="https://bioconductor.org/packages/SingleCellExperiment" class="external-link">SingleCellExperiment</a>
/ <em>Bioconductor</em> ecosystem.</p>
<ul>
<li>[CELLxGENE, <em>Seurat</em> and <em>Bioconductor</em>][C-1]</li>
</ul>
<p>The remaining parts walk through two chapters of <a href="https://bioconductor.org/books/release/OSCA/" class="external-link">OSCA</a>, chosen as
much because I wanted to learn a bit more about these steps as for any
other reason.</p>
<ul>
<li>[Annotating cell types][C-2]</li>
<li>[Differential expression][C-3]</li>
</ul>
<div class="section level3">
<h3 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h3>
<p>This script is derived from the <a href="https://bioconductor.org/books/OSCA.multisample/" class="external-link">OSCA
Multi-Sample</a> book, <a href="https://bioconductor.org/books/release/OSCA.multisample/multi-sample-comparisons.html" class="external-link">Chapter
4</a>: DE analyses between conditions. See the book for full
details.</p>
<p>Initial analysis</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## load data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/MouseGastrulationData" class="external-link">MouseGastrulationData</a></span><span class="op">)</span></span>
<span><span class="va">sce.chimera</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MouseGastrulationData/man/WTChimeraData.html" class="external-link">WTChimeraData</a></span><span class="op">(</span>samples<span class="op">=</span><span class="fl">5</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">sce.chimera</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 29453 20935 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000095742 tomato-td</span></span>
<span><span class="co">#&gt; rowData names(2): ENSEMBL SYMBOL</span></span>
<span><span class="co">#&gt; colnames(20935): cell_9769 cell_9770 ... cell_30702 cell_30703</span></span>
<span><span class="co">#&gt; colData names(11): cell barcode ... doub.density sizeFactor</span></span>
<span><span class="co">#&gt; reducedDimNames(2): pca.corrected.E7.5 pca.corrected.E8.5</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span></span>
<span><span class="co">## feature annotation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce.chimera</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/uniquifyFeatureNames.html" class="external-link">uniquifyFeatureNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce.chimera</span><span class="op">)</span><span class="op">$</span><span class="va">ENSEMBL</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce.chimera</span><span class="op">)</span><span class="op">$</span><span class="va">SYMBOL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## qc</span></span>
<span><span class="va">drop</span> <span class="op">&lt;-</span> <span class="va">sce.chimera</span><span class="op">$</span><span class="va">celltype.mapped</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"stripped"</span>, <span class="st">"Doublet"</span><span class="op">)</span></span>
<span><span class="va">sce.chimera</span> <span class="op">&lt;-</span> <span class="va">sce.chimera</span><span class="op">[</span>,<span class="op">!</span><span class="va">drop</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## normalization</span></span>
<span><span class="va">sce.chimera</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce.chimera</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## variance modeling</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="va">dec.chimera</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce.chimera</span>, block<span class="op">=</span><span class="va">sce.chimera</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span><span class="va">chosen.hvgs</span> <span class="op">&lt;-</span> <span class="va">dec.chimera</span><span class="op">$</span><span class="va">bio</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span></span>
<span><span class="co">## merge batches</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">01001001</span><span class="op">)</span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/correctExperiments.html" class="external-link">correctExperiments</a></span><span class="op">(</span></span>
<span>    <span class="va">sce.chimera</span>, </span>
<span>    batch<span class="op">=</span><span class="va">sce.chimera</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>    subset.row<span class="op">=</span><span class="va">chosen.hvgs</span>,</span>
<span>    PARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/BatchelorParam.html" class="external-link">FastMnnParam</a></span><span class="op">(</span></span>
<span>        merge.order<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">5</span><span class="op">)</span>, <span class="co"># WT (3 replicates)</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">6</span><span class="op">)</span>  <span class="co"># td-Tomato (3 replicates)</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">## cluster</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span><span class="va">merged</span>, use.dimred<span class="op">=</span><span class="st">"corrected"</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_louvain.html" class="external-link">cluster_louvain</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">merged</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">$</span><span class="va">membership</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## dimensionality reduction</span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">merged</span>, dimred<span class="op">=</span><span class="st">"corrected"</span>, external_neighbors<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">merged</span>, dimred<span class="op">=</span><span class="st">"corrected"</span>, external_neighbors<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>This document was produced with the following <em>R</em>
software:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] batchelor_1.17.0             scran_1.29.0                </span></span>
<span><span class="co">#&gt;  [3] scater_1.29.0                ggplot2_3.4.2               </span></span>
<span><span class="co">#&gt;  [5] scuttle_1.11.0               MouseGastrulationData_1.15.0</span></span>
<span><span class="co">#&gt;  [7] SpatialExperiment_1.11.0     SingleCellExperiment_1.23.0 </span></span>
<span><span class="co">#&gt;  [9] SummarizedExperiment_1.31.1  Biobase_2.61.0              </span></span>
<span><span class="co">#&gt; [11] GenomicRanges_1.53.1         GenomeInfoDb_1.37.1         </span></span>
<span><span class="co">#&gt; [13] IRanges_2.35.1               S4Vectors_0.39.1            </span></span>
<span><span class="co">#&gt; [15] BiocGenerics_0.47.0          MatrixGenerics_1.13.0       </span></span>
<span><span class="co">#&gt; [17] matrixStats_1.0.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] jsonlite_1.8.5                magrittr_2.0.3               </span></span>
<span><span class="co">#&gt;   [3] ggbeeswarm_0.7.2              magick_2.7.4                 </span></span>
<span><span class="co">#&gt;   [5] rmarkdown_2.22                fs_1.6.2                     </span></span>
<span><span class="co">#&gt;   [7] zlibbioc_1.47.0               ragg_1.2.5                   </span></span>
<span><span class="co">#&gt;   [9] vctrs_0.6.2                   memoise_2.0.1                </span></span>
<span><span class="co">#&gt;  [11] DelayedMatrixStats_1.23.0     RCurl_1.98-1.12              </span></span>
<span><span class="co">#&gt;  [13] htmltools_0.5.5               S4Arrays_1.1.4               </span></span>
<span><span class="co">#&gt;  [15] AnnotationHub_3.9.1           curl_5.0.1                   </span></span>
<span><span class="co">#&gt;  [17] BiocNeighbors_1.19.0          Rhdf5lib_1.23.0              </span></span>
<span><span class="co">#&gt;  [19] SparseArray_1.1.10            rhdf5_2.45.0                 </span></span>
<span><span class="co">#&gt;  [21] sass_0.4.6                    bslib_0.5.0                  </span></span>
<span><span class="co">#&gt;  [23] desc_1.4.2                    cachem_1.0.8                 </span></span>
<span><span class="co">#&gt;  [25] ResidualMatrix_1.11.0         igraph_1.4.3                 </span></span>
<span><span class="co">#&gt;  [27] mime_0.12                     lifecycle_1.0.3              </span></span>
<span><span class="co">#&gt;  [29] pkgconfig_2.0.3               rsvd_1.0.5                   </span></span>
<span><span class="co">#&gt;  [31] Matrix_1.5-4.1                R6_2.5.1                     </span></span>
<span><span class="co">#&gt;  [33] fastmap_1.1.1                 GenomeInfoDbData_1.2.10      </span></span>
<span><span class="co">#&gt;  [35] shiny_1.7.4                   digest_0.6.31                </span></span>
<span><span class="co">#&gt;  [37] colorspace_2.1-0              AnnotationDbi_1.63.1         </span></span>
<span><span class="co">#&gt;  [39] rprojroot_2.0.3               dqrng_0.3.0                  </span></span>
<span><span class="co">#&gt;  [41] irlba_2.3.5.1                 ExperimentHub_2.7.1          </span></span>
<span><span class="co">#&gt;  [43] textshaping_0.3.6             RSQLite_2.3.1                </span></span>
<span><span class="co">#&gt;  [45] beachmat_2.17.8               filelock_1.0.2               </span></span>
<span><span class="co">#&gt;  [47] fansi_1.0.4                   httr_1.4.6                   </span></span>
<span><span class="co">#&gt;  [49] compiler_4.3.0                bit64_4.0.5                  </span></span>
<span><span class="co">#&gt;  [51] withr_2.5.0                   BiocParallel_1.35.2          </span></span>
<span><span class="co">#&gt;  [53] viridis_0.6.3                 DBI_1.1.3                    </span></span>
<span><span class="co">#&gt;  [55] HDF5Array_1.29.3              R.utils_2.12.2               </span></span>
<span><span class="co">#&gt;  [57] rappdirs_0.3.3                DelayedArray_0.27.5          </span></span>
<span><span class="co">#&gt;  [59] bluster_1.11.1                rjson_0.2.21                 </span></span>
<span><span class="co">#&gt;  [61] tools_4.3.0                   vipor_0.4.5                  </span></span>
<span><span class="co">#&gt;  [63] beeswarm_0.4.0                interactiveDisplayBase_1.39.0</span></span>
<span><span class="co">#&gt;  [65] httpuv_1.6.11                 R.oo_1.25.0                  </span></span>
<span><span class="co">#&gt;  [67] glue_1.6.2                    rhdf5filters_1.13.2          </span></span>
<span><span class="co">#&gt;  [69] promises_1.2.0.1              grid_4.3.0                   </span></span>
<span><span class="co">#&gt;  [71] Rtsne_0.16                    cluster_2.1.4                </span></span>
<span><span class="co">#&gt;  [73] generics_0.1.3                gtable_0.3.3                 </span></span>
<span><span class="co">#&gt;  [75] R.methodsS3_1.8.2             metapod_1.9.0                </span></span>
<span><span class="co">#&gt;  [77] BiocSingular_1.17.0           ScaledMatrix_1.9.1           </span></span>
<span><span class="co">#&gt;  [79] utf8_1.2.3                    XVector_0.41.1               </span></span>
<span><span class="co">#&gt;  [81] ggrepel_0.9.3                 BiocVersion_3.18.0           </span></span>
<span><span class="co">#&gt;  [83] pillar_1.9.0                  stringr_1.5.0                </span></span>
<span><span class="co">#&gt;  [85] limma_3.57.4                  BumpyMatrix_1.9.0            </span></span>
<span><span class="co">#&gt;  [87] later_1.3.1                   dplyr_1.1.2                  </span></span>
<span><span class="co">#&gt;  [89] BiocFileCache_2.9.0           lattice_0.21-8               </span></span>
<span><span class="co">#&gt;  [91] bit_4.0.5                     tidyselect_1.2.0             </span></span>
<span><span class="co">#&gt;  [93] locfit_1.5-9.8                Biostrings_2.69.1            </span></span>
<span><span class="co">#&gt;  [95] knitr_1.43                    gridExtra_2.3                </span></span>
<span><span class="co">#&gt;  [97] edgeR_3.43.4                  xfun_0.39                    </span></span>
<span><span class="co">#&gt;  [99] statmod_1.5.0                 DropletUtils_1.21.0          </span></span>
<span><span class="co">#&gt; [101] stringi_1.7.12                yaml_2.3.7                   </span></span>
<span><span class="co">#&gt; [103] evaluate_0.21                 codetools_0.2-19             </span></span>
<span><span class="co">#&gt; [105] tibble_3.2.1                  BiocManager_1.30.21          </span></span>
<span><span class="co">#&gt; [107] cli_3.6.1                     uwot_0.1.14                  </span></span>
<span><span class="co">#&gt; [109] xtable_1.8-4                  systemfonts_1.0.4            </span></span>
<span><span class="co">#&gt; [111] munsell_0.5.0                 jquerylib_0.1.4              </span></span>
<span><span class="co">#&gt; [113] Rcpp_1.0.10                   dbplyr_2.3.2                 </span></span>
<span><span class="co">#&gt; [115] png_0.1-8                     parallel_4.3.0               </span></span>
<span><span class="co">#&gt; [117] ellipsis_0.3.2                pkgdown_2.0.7                </span></span>
<span><span class="co">#&gt; [119] blob_1.2.4                    sparseMatrixStats_1.13.0     </span></span>
<span><span class="co">#&gt; [121] bitops_1.0-7                  viridisLite_0.4.2            </span></span>
<span><span class="co">#&gt; [123] scales_1.2.1                  purrr_1.0.1                  </span></span>
<span><span class="co">#&gt; [125] crayon_1.5.2                  rlang_1.1.1                  </span></span>
<span><span class="co">#&gt; [127] KEGGREST_1.41.0</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Martin Morgan, NIH NHGRI U24HG004059.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
