<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="XM2023">
<title>D. From CELLxGENE to R and Bioconductor • XM2023</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="D. From CELLxGENE to R and Bioconductor">
<meta property="og:description" content="XM2023">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">XM2023</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/a_talk.html">A. How Bioconductor Advances Science</a>
    <a class="dropdown-item" href="../articles/b_course_part_1.html">B. Using R to Understand Bioinformatic Results</a>
    <a class="dropdown-item" href="../articles/c_course_part_2.html">C. R and Bioconductor for Genomic Analysis</a>
    <a class="dropdown-item" href="../articles/d_cxg.html">D. From CELLxGENE to R and Bioconductor</a>
    <a class="dropdown-item" href="../articles/e_sc_workflows.html">E. Seurat and Bioconductor Single-Cell Sequence Analysis Workflows</a>
    <a class="dropdown-item" href="../articles/f_osca_cell_annotation.html">F. Annotating cell types in Bioconductor</a>
    <a class="dropdown-item" href="../articles/g_osca_differential_expression.html">G. Differential Expression in Bioconductor</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mtmorgan/XM2023">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>D. From CELLxGENE to R and Bioconductor</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mtmorgan/XM2023/blob/HEAD/vignettes/d_cxg.Rmd" class="external-link"><code>vignettes/d_cxg.Rmd</code></a></small>
      <div class="d-none name"><code>d_cxg.Rmd</code></div>
    </div>

    
    
<p>We use the following packages during this section of the
workshop.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## package developed for this workshop</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://mtmorgan.github.io/XM2023/">XM2023</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## general programming tools</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## data access</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">cellxgenedp</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## single cell data representation in R</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span> <span class="co"># Bioconductor representation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>               <span class="co"># Seurat representation</span></span></code></pre></div>
<div class="section level2">
<h2 id="data-transformations">Data Transformations<a class="anchor" aria-label="anchor" href="#data-transformations"></a>
</h2>
<p>FASTQ files</p>
<ul>
<li>DNA sequences and quality scores</li>
<li>Very large</li>
<li>Processed e.g., by CellRanger or other software</li>
<li>Quality control, summary to count matrix (below)</li>
</ul>
<p>Count matrix</p>
<ul>
<li>Usually genes (rows) x cells (columns)</li>
<li>Can be large (e.g., 30,000 genes x 50,000 cells)</li>
<li>Usually <em>very</em> sparse, e.g., 95% of cells ‘0’.</li>
<li>Still large enough to require a decent amount of computing power,
e.g., 32 GB RAM, 8 CPU for some steps.</li>
</ul>
<p>Common formats for count matrix data</p>
<ul>
<li>CSV file – lots of zero’s so very wasteful of space.</li>
<li>‘Matrix Market’ sparse matrix files, e.g., tuples of &lt;row,
column, count&gt; for non-zero values.</li>
<li>HDF5, e.g., <a href="https://linnarssonlab.org/loompy/format/index.html" class="external-link">.loom</a> or
<code>.h5ad</code> (<a href="https://anndata.readthedocs.io/en/latest/index.html" class="external-link">anndata</a>).</li>
<li>RDS – An <em>R</em> file, in <em>CELLxGENE</em> these are
<em>Seurat</em> objects but in general they could contain any <em>R</em>
object.</li>
</ul>
<p>Matrix representation in <em>R</em></p>
<ul>
<li>In-memory sparse matrices: <code>dgCMatrix</code> class from the
Matrix package</li>
<li>On-disk representation via <em>Bioconductor</em>’s <a href="https://bioconductor.org/packages/DelayedArray" class="external-link">DelayedArray</a> /
<a href="https://bioconductor.org/packages/HDF5Array" class="external-link">HDF5Array</a>.</li>
</ul>
<p>Counts &amp; annotations</p>
<ul>
<li>CSV and Matrix Market files store just counts; usually annotations
on columns (e.g., what sample did each cell come from?) are stored
separately.</li>
<li>HDF5 file formats coordinate row and column annotations with count
data.</li>
<li>
<em>R</em> software tries to offer a coordinated representation of
counts and metadata, e.g., the <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a> or <em>Bioconductor</em>
<a href="https://bioconductor.org/packages/SingleCellExperiment" class="external-link">SingleCellExperiment</a>
objects.</li>
</ul>
</div>
<div class="section level2">
<h2 id="cellxgene">CELLxGENE<a class="anchor" aria-label="anchor" href="#cellxgene"></a>
</h2>
<p><a href="https://cellxgene.cziscience.com/datasets" class="external-link">Data
Portal</a></p>
<p><img src="images/CXGDP.png"></p>
<p>What’s available?</p>
<ul>
<li>Collections and datasets contributed by the single-cell community,
with some overlap with data sets in the HCA Data Portal.</li>
<li>FASTQ files</li>
<li>
<code>.h5ad</code> or <code>.RDS</code> summarized count data and
cell metadata, as well as reduced-dimension (e.g., UMAP) representations
<ul>
<li>Summarized count files provided by the contributor / individual lab,
so of uncertain provenance</li>
</ul>
</li>
<li>Easy to download count data</li>
<li>Easy to visualize (!)</li>
</ul>
<p><img src="images/CXGViz.png"></p>
<div class="section level3">
<h3 id="programatic-discovery">Programatic Discovery<a class="anchor" aria-label="anchor" href="#programatic-discovery"></a>
</h3>
<p>Why use an <em>R</em> script when the Data Portals exist?</p>
<ul>
<li>Easily reproducible</li>
<li>Flexible exploration of rich &amp; complex data</li>
<li>Direct integration with <em>Seurat</em> or <em>Bioconductor</em>
single-cell workflows</li>
</ul>
<p>Load the <a href="https://bioconductor.org/packages/cellxgenedp" class="external-link">cellxgenedp</a>
package</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">cellxgenedp</span><span class="op">)</span></span></code></pre></div>
<p>Retrieve the current database, and use ‘tidy’ functionality to mimic
the graphical selection in the web browser – 10x 3’ v3 (EFO:0009922)
assay, African American ethnicity, female gender)</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">db</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/cellxgenedp/man/db.html" class="external-link">db</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">african_american_female</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/cellxgenedp/man/query.html" class="external-link">datasets</a></span><span class="op">(</span><span class="va">db</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/cellxgenedp/man/facets.html" class="external-link">facets_filter</a></span><span class="op">(</span><span class="va">assay</span>, <span class="st">"ontology_term_id"</span>, <span class="st">"EFO:0009922"</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/cellxgenedp/man/facets.html" class="external-link">facets_filter</a></span><span class="op">(</span><span class="va">self_reported_ethnicity</span>, <span class="st">"label"</span>, <span class="st">"African American"</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/cellxgenedp/man/facets.html" class="external-link">facets_filter</a></span><span class="op">(</span><span class="va">sex</span>, <span class="st">"label"</span>, <span class="st">"female"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">african_american_female</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 32 × 26</span></span></span>
<span><span class="co">#&gt;    dataset_id                 collection_id donor_id assay  cell_count cell_type</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                      <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 8693bc2c-d84f-450f-983a-d… bcb61471-2a4… <span style="color: #949494;">&lt;list&gt;</span>   <span style="color: #949494;">&lt;list&gt;</span>     <span style="text-decoration: underline;">304</span>652 <span style="color: #949494;">&lt;list&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 37f30005-8a5c-4bdf-88aa-a… bcb61471-2a4… <span style="color: #949494;">&lt;list&gt;</span>   <span style="color: #949494;">&lt;list&gt;</span>     <span style="text-decoration: underline;">172</span>847 <span style="color: #949494;">&lt;list&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 54914718-c0d3-4e17-9e06-d… bcb61471-2a4… <span style="color: #949494;">&lt;list&gt;</span>   <span style="color: #949494;">&lt;list&gt;</span>     <span style="text-decoration: underline;">107</span>344 <span style="color: #949494;">&lt;list&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 4c4dedb7-1f74-4be7-9916-2… b953c942-f5d… <span style="color: #949494;">&lt;list&gt;</span>   <span style="color: #949494;">&lt;list&gt;</span>      <span style="text-decoration: underline;">23</span>197 <span style="color: #949494;">&lt;list&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 8716534a-4c1c-414c-aa98-4… b953c942-f5d… <span style="color: #949494;">&lt;list&gt;</span>   <span style="color: #949494;">&lt;list&gt;</span>      <span style="text-decoration: underline;">96</span>129 <span style="color: #949494;">&lt;list&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 24205601-0780-4bf2-b1d9-0… c9706a92-0e5… <span style="color: #949494;">&lt;list&gt;</span>   <span style="color: #949494;">&lt;list&gt;</span>      <span style="text-decoration: underline;">31</span>696 <span style="color: #949494;">&lt;list&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> a963dd1b-ee7e-42b3-8afe-0… a98b828a-622… <span style="color: #949494;">&lt;list&gt;</span>   <span style="color: #949494;">&lt;list&gt;</span>      <span style="text-decoration: underline;">48</span>783 <span style="color: #949494;">&lt;list&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 36e76662-60b8-4193-8a70-1… 625f6bf4-2f3… <span style="color: #949494;">&lt;list&gt;</span>   <span style="color: #949494;">&lt;list&gt;</span>      <span style="text-decoration: underline;">46</span>500 <span style="color: #949494;">&lt;list&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 611c3577-81cb-4795-ad3a-4… b9fc3d70-5a7… <span style="color: #949494;">&lt;list&gt;</span>   <span style="color: #949494;">&lt;list&gt;</span>      <span style="text-decoration: underline;">49</span>139 <span style="color: #949494;">&lt;list&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 0ed93577-0ff5-435e-b599-b… b9fc3d70-5a7… <span style="color: #949494;">&lt;list&gt;</span>   <span style="color: #949494;">&lt;list&gt;</span>     <span style="text-decoration: underline;">109</span>995 <span style="color: #949494;">&lt;list&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 22 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 20 more variables: dataset_deployments &lt;chr&gt;, development_stage &lt;list&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   disease &lt;list&gt;, is_primary_data &lt;chr&gt;, is_valid &lt;lgl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   mean_genes_per_cell &lt;dbl&gt;, name &lt;chr&gt;, organism &lt;list&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   processing_status &lt;list&gt;, published &lt;lgl&gt;, revision &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   schema_version &lt;chr&gt;, self_reported_ethnicity &lt;list&gt;, sex &lt;list&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   suspension_type &lt;list&gt;, tissue &lt;list&gt;, tombstone &lt;lgl&gt;, …</span></span></span></code></pre></div>
<p>A particular collection is</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">collection_id</span> <span class="op">&lt;-</span> <span class="st">"c9706a92-0e5f-46c1-96d8-20e42467f287"</span></span>
<span><span class="va">african_american_female</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">collection_id</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">.env</span><span class="op">$</span><span class="va">collection_id</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 26</span></span></span>
<span><span class="co">#&gt;   dataset_id                  collection_id donor_id assay  cell_count cell_type</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;list&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 24205601-0780-4bf2-b1d9-0e… c9706a92-0e5… <span style="color: #949494;">&lt;list&gt;</span>   <span style="color: #949494;">&lt;list&gt;</span>      <span style="text-decoration: underline;">31</span>696 <span style="color: #949494;">&lt;list&gt;</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 20 more variables: dataset_deployments &lt;chr&gt;, development_stage &lt;list&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   disease &lt;list&gt;, is_primary_data &lt;chr&gt;, is_valid &lt;lgl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   mean_genes_per_cell &lt;dbl&gt;, name &lt;chr&gt;, organism &lt;list&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   processing_status &lt;list&gt;, published &lt;lgl&gt;, revision &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   schema_version &lt;chr&gt;, self_reported_ethnicity &lt;list&gt;, sex &lt;list&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   suspension_type &lt;list&gt;, tissue &lt;list&gt;, tombstone &lt;lgl&gt;, created_at &lt;date&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   published_at &lt;date&gt;, updated_at &lt;date&gt;</span></span></span></code></pre></div>
<p>The dataset associated with this collection is</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataset</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">african_american_female</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">collection_id</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">.env</span><span class="op">$</span><span class="va">collection_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">dataset_id</span><span class="op">)</span></span>
<span><span class="va">dataset_id</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">dataset</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">dataset_id</span><span class="op">)</span></span>
<span><span class="va">dataset_id</span></span>
<span><span class="co">#&gt; [1] "24205601-0780-4bf2-b1d9-0e3cacbc2cd6"</span></span></code></pre></div>
<p>The helper function <code><a href="../reference/xm2023_cxg_dataset.html">xm2023_cxg_dataset()</a></code> provides a
summary of this dataset.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/xm2023_cxg_dataset.html">xm2023_cxg_dataset</a></span><span class="op">(</span><span class="va">dataset_id</span><span class="op">)</span></span>
<span><span class="co">#&gt; title: A single-cell atlas of the healthy breast tissues reveals</span></span>
<span><span class="co">#&gt;     clinically relevant clusters of breast epithelial cells</span></span>
<span><span class="co">#&gt; description: Single-cell RNA sequencing (scRNA-seq) is an evolving</span></span>
<span><span class="co">#&gt;     technology used to elucidate the cellular architecture of adult</span></span>
<span><span class="co">#&gt;     organs. Previous scRNA-seq on breast tissue utilized reduction</span></span>
<span><span class="co">#&gt;     mammoplasty samples, which are often histologically abnormal. We</span></span>
<span><span class="co">#&gt;     report a rapid tissue collection/processing protocol to perform</span></span>
<span><span class="co">#&gt;     scRNA-seq of breast biopsies of healthy women and identify 23</span></span>
<span><span class="co">#&gt;     breast epithelial cell clusters. Putative cell-of-origin signatures</span></span>
<span><span class="co">#&gt;     derived from these clusters are applied to analyze transcriptomes</span></span>
<span><span class="co">#&gt;     of ~3,000 breast cancers. Gene signatures derived from mature</span></span>
<span><span class="co">#&gt;     luminal cell clusters are enriched in ~68% of breast cancers,</span></span>
<span><span class="co">#&gt;     whereas a signature from a luminal progenitor cluster is enriched</span></span>
<span><span class="co">#&gt;     in ~20% of breast cancers. Overexpression of luminal progenitor</span></span>
<span><span class="co">#&gt;     cluster-derived signatures in HER2+, but not in other subtypes, is</span></span>
<span><span class="co">#&gt;     associated with unfavorable outcome. We identify TBX3 and PDK4 as</span></span>
<span><span class="co">#&gt;     genes co-expressed with estrogen receptor (ER) in the normal</span></span>
<span><span class="co">#&gt;     breasts, and their expression analyses in &gt;550 breast cancers</span></span>
<span><span class="co">#&gt;     enable prognostically relevant subclassification of ER+ breast</span></span>
<span><span class="co">#&gt;     cancers.</span></span>
<span><span class="co">#&gt; authors: Bhat-Nakshatri, Poornima; Gao, Hongyu; Sheng, Liu; McGuire,</span></span>
<span><span class="co">#&gt;     Patrick C.; Xuei, Xiaoling; Wan, Jun; Liu, Yunlong; Althouse,</span></span>
<span><span class="co">#&gt;     Sandra K.; Colter, Austyn; Sandusky, George; Storniolo, Anna Maria;</span></span>
<span><span class="co">#&gt;     Nakshatri, Harikrishna</span></span>
<span><span class="co">#&gt; journal: Cell Reports Medicine</span></span>
<span><span class="co">#&gt; assays: 10x 3' v2; 10x 3' v3</span></span>
<span><span class="co">#&gt; organism: Homo sapiens</span></span>
<span><span class="co">#&gt; ethnicity: African American; Chinese; European</span></span></code></pre></div>
<p>‘Join’ selected datasets and files to identify the files associated
with these datasets.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_files</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span></span>
<span>        <span class="va">dataset</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/cellxgenedp/man/query.html" class="external-link">files</a></span><span class="op">(</span><span class="va">db</span><span class="op">)</span>,</span>
<span>        by <span class="op">=</span> <span class="st">"dataset_id"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="va">selected_files</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4 × 8</span></span></span>
<span><span class="co">#&gt;   dataset_id          file_id filename filetype s3_uri user_submitted created_at</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;date&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 24205601-0780-4bf2… e1c842… <span style="color: #949494;">"</span>raw.h5… RAW_H5AD s3://… TRUE           1970-01-01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 24205601-0780-4bf2… 15e9d9… <span style="color: #949494;">"</span>local.… H5AD     s3://… TRUE           1970-01-01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 24205601-0780-4bf2… 0d3974… <span style="color: #949494;">""</span>       CXG      s3://… TRUE           1970-01-01</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 24205601-0780-4bf2… e254f9… <span style="color: #949494;">"</span>local.… RDS      s3://… TRUE           1970-01-01</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: updated_at &lt;date&gt;</span></span></span></code></pre></div>
<p>Select the first ‘CXG’ file available in this subset of data; for
reproducibility we retrieve the dataset id…</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_files</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">filetype</span> <span class="op">==</span> <span class="st">"CXG"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 1 × 8</span></span></span>
<span><span class="co">#&gt;   dataset_id          file_id filename filetype s3_uri user_submitted created_at</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;lgl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;date&gt;</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 24205601-0780-4bf2… 0d3974… <span style="color: #949494;">""</span>       CXG      s3://… TRUE           1970-01-01</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: updated_at &lt;date&gt;</span></span></span></code></pre></div>
<p>Visualize this ‘CXG’ file in the browser…</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_files</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">filetype</span> <span class="op">==</span> <span class="st">"CXG"</span>, <span class="va">dataset_id</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">.env</span><span class="op">$</span><span class="va">dataset_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/cellxgenedp/man/query.html" class="external-link">datasets_visualize</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>…or select the ‘H5AD’ or ‘RDS’ (Seurat) file associated with the
dataset and download it for subsequent processing in <em>R</em></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_file</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">selected_files</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">filetype</span> <span class="op">==</span> <span class="st">"RDS"</span>, <span class="va">dataset_id</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">.env</span><span class="op">$</span><span class="va">dataset_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/cellxgenedp/man/query.html" class="external-link">files_download</a></span><span class="op">(</span>dry.run <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">h5ad_file</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">selected_files</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">filetype</span> <span class="op">==</span> <span class="st">"H5AD"</span>, <span class="va">dataset_id</span> <span class="op">==</span> <span class="va">.env</span><span class="op">$</span><span class="va">dataset_id</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/cellxgenedp/man/query.html" class="external-link">files_download</a></span><span class="op">(</span>dry.run <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>The downloaded file is cached, so the next time access is fast.</p>
</div>
</div>
<div class="section level2">
<h2 id="seurat">
<em>Seurat</em><a class="anchor" aria-label="anchor" href="#seurat"></a>
</h2>
<div class="section level3">
<h3 id="data-input-representation">Data input &amp; representation<a class="anchor" aria-label="anchor" href="#data-input-representation"></a>
</h3>
<p>Fortunately, CELLxGENE distributes Seurat (v. 4) files, and they can
be input directly.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="va">seurat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="va">seurat_file</span><span class="op">)</span></span>
<span><span class="va">seurat</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 33234 features across 31696 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (33234 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  3 dimensional reductions calculated: pca, tsne, umap</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="standard-workflow">Standard workflow<a class="anchor" aria-label="anchor" href="#standard-workflow"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="bioconductor">
<em>Bioconductor</em><a class="anchor" aria-label="anchor" href="#bioconductor"></a>
</h2>
<div class="section level3">
<h3 id="data-input-representation-1">Data input &amp; representation<a class="anchor" aria-label="anchor" href="#data-input-representation-1"></a>
</h3>
<p>The <code>SingleCellExperiment</code> is used to represent
‘rectangular’ single cell expression and other data in <em>R</em> /
<em>Bioconductor</em>. It coordinates a gene x cell count matricies
(<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay()</a></code>) with annotations on the genes
(<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code>) and columns (<code>cellData()</code>), and with
reduced-dimension summaries.</p>
<p>An effective way to represent the <code>.h5ad</code> data as a
<code>SingleCellExperiment</code> is using
<code><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html" class="external-link">zellkonverter::readH5AD()</a></code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h5ad</span> <span class="op">&lt;-</span> <span class="fu">zellkonverter</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html" class="external-link">readH5AD</a></span><span class="op">(</span><span class="va">h5ad_file</span>, reader <span class="op">=</span> <span class="st">"R"</span>, use_hdf5 <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Registered S3 methods overwritten by 'zellkonverter':</span></span>
<span><span class="co">#&gt;   method                                             from      </span></span>
<span><span class="co">#&gt;   py_to_r.numpy.ndarray                              reticulate</span></span>
<span><span class="co">#&gt;   py_to_r.pandas.core.arrays.categorical.Categorical reticulate</span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span>
<span></span>
<span><span class="co">#&gt; Warning in H5Aread(A, ...): Reading attribute data of type 'ENUM' not yet</span></span>
<span><span class="co">#&gt; implemented. Values replaced by NA's.</span></span></code></pre></div>
<p>Displaying the object…</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h5ad</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 33234 31696 </span></span>
<span><span class="co">#&gt; metadata(3): default_embedding schema_version title</span></span>
<span><span class="co">#&gt; assays(1): X</span></span>
<span><span class="co">#&gt; rownames(33234): ENSG00000243485 ENSG00000237613 ... ENSG00000277475</span></span>
<span><span class="co">#&gt;   ENSG00000268674</span></span>
<span><span class="co">#&gt; rowData names(4): feature_is_filtered feature_name feature_reference</span></span>
<span><span class="co">#&gt;   feature_biotype</span></span>
<span><span class="co">#&gt; colnames(31696): CMGpool_AAACCCAAGGACAACC CMGpool_AAACCCACAATCTCTT ...</span></span>
<span><span class="co">#&gt;   K109064_TTTGTTGGTTGCATCA K109064_TTTGTTGGTTGGACCC</span></span>
<span><span class="co">#&gt; colData names(34): donor_id self_reported_ethnicity_ontology_term_id</span></span>
<span><span class="co">#&gt;   ... self_reported_ethnicity development_stage</span></span>
<span><span class="co">#&gt; reducedDimNames(3): X_pca X_tsne X_umap</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
<p>…suggests the data available and how to access it – there are 33234
genes and 31696 cells. The ‘raw’ data include counts
<code>assays(h5ad, "counts")</code>, annotations on each gene
(<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code>) and cell (<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>), etc…</p>
<p>Working with <code>SingleCellExperiment</code> objects is described
in additional detail subsequent articles.</p>
</div>
<div class="section level3">
<h3 id="standard-workflow-1">Standard workflow<a class="anchor" aria-label="anchor" href="#standard-workflow-1"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>This document was produced with the following <em>R</em>
software:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] SeuratObject_4.1.3          Seurat_4.3.0               </span></span>
<span><span class="co">#&gt;  [3] SingleCellExperiment_1.23.0 SummarizedExperiment_1.31.1</span></span>
<span><span class="co">#&gt;  [5] Biobase_2.61.0              GenomicRanges_1.53.1       </span></span>
<span><span class="co">#&gt;  [7] GenomeInfoDb_1.37.1         IRanges_2.35.1             </span></span>
<span><span class="co">#&gt;  [9] S4Vectors_0.39.1            BiocGenerics_0.47.0        </span></span>
<span><span class="co">#&gt; [11] MatrixGenerics_1.13.0       matrixStats_1.0.0          </span></span>
<span><span class="co">#&gt; [13] cellxgenedp_1.5.0           dplyr_1.1.2                </span></span>
<span><span class="co">#&gt; [15] XM2023_0.0.1               </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3      jsonlite_1.8.5          magrittr_2.0.3         </span></span>
<span><span class="co">#&gt;   [4] spatstat.utils_3.0-3    rmarkdown_2.22          fs_1.6.2               </span></span>
<span><span class="co">#&gt;   [7] zlibbioc_1.47.0         ragg_1.2.5              vctrs_0.6.3            </span></span>
<span><span class="co">#&gt;  [10] ROCR_1.0-11             spatstat.explore_3.2-1  memoise_2.0.1          </span></span>
<span><span class="co">#&gt;  [13] RCurl_1.98-1.12         htmltools_0.5.5         S4Arrays_1.1.4         </span></span>
<span><span class="co">#&gt;  [16] curl_5.0.1              Rhdf5lib_1.23.0         rhdf5_2.45.0           </span></span>
<span><span class="co">#&gt;  [19] SparseArray_1.1.10      sctransform_0.3.5       sass_0.4.6             </span></span>
<span><span class="co">#&gt;  [22] parallelly_1.36.0       KernSmooth_2.23-21      bslib_0.5.0            </span></span>
<span><span class="co">#&gt;  [25] basilisk_1.13.1         htmlwidgets_1.6.2       desc_1.4.2             </span></span>
<span><span class="co">#&gt;  [28] ica_1.0-3               plyr_1.8.8              plotly_4.10.2          </span></span>
<span><span class="co">#&gt;  [31] zoo_1.8-12              cachem_1.0.8            igraph_1.5.0           </span></span>
<span><span class="co">#&gt;  [34] mime_0.12               lifecycle_1.0.3         pkgconfig_2.0.3        </span></span>
<span><span class="co">#&gt;  [37] Matrix_1.5-4.1          R6_2.5.1                fastmap_1.1.1          </span></span>
<span><span class="co">#&gt;  [40] GenomeInfoDbData_1.2.10 fitdistrplus_1.1-11     future_1.32.0          </span></span>
<span><span class="co">#&gt;  [43] shiny_1.7.4             digest_0.6.31           colorspace_2.1-0       </span></span>
<span><span class="co">#&gt;  [46] patchwork_1.1.2         tensor_1.5              rprojroot_2.0.3        </span></span>
<span><span class="co">#&gt;  [49] irlba_2.3.5.1           textshaping_0.3.6       filelock_1.0.2         </span></span>
<span><span class="co">#&gt;  [52] progressr_0.13.0        spatstat.sparse_3.0-1   fansi_1.0.4            </span></span>
<span><span class="co">#&gt;  [55] polyclip_1.10-4         abind_1.4-5             httr_1.4.6             </span></span>
<span><span class="co">#&gt;  [58] compiler_4.3.0          withr_2.5.0             rjsoncons_1.0.0        </span></span>
<span><span class="co">#&gt;  [61] HDF5Array_1.29.3        MASS_7.3-60             DelayedArray_0.27.5    </span></span>
<span><span class="co">#&gt;  [64] tools_4.3.0             lmtest_0.9-40           httpuv_1.6.11          </span></span>
<span><span class="co">#&gt;  [67] future.apply_1.11.0     goftest_1.2-3           glue_1.6.2             </span></span>
<span><span class="co">#&gt;  [70] rhdf5filters_1.13.3     nlme_3.1-162            promises_1.2.0.1       </span></span>
<span><span class="co">#&gt;  [73] grid_4.3.0              Rtsne_0.16              reshape2_1.4.4         </span></span>
<span><span class="co">#&gt;  [76] cluster_2.1.4           generics_0.1.3          spatstat.data_3.0-1    </span></span>
<span><span class="co">#&gt;  [79] gtable_0.3.3            tidyr_1.3.0             data.table_1.14.8      </span></span>
<span><span class="co">#&gt;  [82] sp_1.6-1                utf8_1.2.3              XVector_0.41.1         </span></span>
<span><span class="co">#&gt;  [85] spatstat.geom_3.2-1     RcppAnnoy_0.0.20        ggrepel_0.9.3          </span></span>
<span><span class="co">#&gt;  [88] RANN_2.6.1              pillar_1.9.0            stringr_1.5.0          </span></span>
<span><span class="co">#&gt;  [91] later_1.3.1             splines_4.3.0           lattice_0.21-8         </span></span>
<span><span class="co">#&gt;  [94] deldir_1.0-9            survival_3.5-5          tidyselect_1.2.0       </span></span>
<span><span class="co">#&gt;  [97] miniUI_0.1.1.1          pbapply_1.7-0           knitr_1.43             </span></span>
<span><span class="co">#&gt; [100] gridExtra_2.3           scattermore_1.2         xfun_0.39              </span></span>
<span><span class="co">#&gt; [103] DT_0.28                 stringi_1.7.12          lazyeval_0.2.2         </span></span>
<span><span class="co">#&gt; [106] yaml_2.3.7              evaluate_0.21           codetools_0.2-19       </span></span>
<span><span class="co">#&gt; [109] tibble_3.2.1            cli_3.6.1               uwot_0.1.14            </span></span>
<span><span class="co">#&gt; [112] xtable_1.8-4            reticulate_1.30         systemfonts_1.0.4      </span></span>
<span><span class="co">#&gt; [115] munsell_0.5.0           jquerylib_0.1.4         zellkonverter_1.11.1   </span></span>
<span><span class="co">#&gt; [118] Rcpp_1.0.10             dir.expiry_1.9.0        spatstat.random_3.1-5  </span></span>
<span><span class="co">#&gt; [121] globals_0.16.2          png_0.1-8               parallel_4.3.0         </span></span>
<span><span class="co">#&gt; [124] ellipsis_0.3.2          pkgdown_2.0.7           ggplot2_3.4.2          </span></span>
<span><span class="co">#&gt; [127] basilisk.utils_1.13.1   bitops_1.0-7            listenv_0.9.0          </span></span>
<span><span class="co">#&gt; [130] viridisLite_0.4.2       scales_1.2.1            ggridges_0.5.4         </span></span>
<span><span class="co">#&gt; [133] leiden_0.4.3            purrr_1.0.1             crayon_1.5.2           </span></span>
<span><span class="co">#&gt; [136] rlang_1.1.1             cowplot_1.1.1</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Martin Morgan, NIH NHGRI U24HG004059.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
