<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="XM2023">
<title>F. Annotating cell types in Bioconductor • XM2023</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="F. Annotating cell types in Bioconductor">
<meta property="og:description" content="XM2023">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">XM2023</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9012</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/a_talk.html">A. How Bioconductor Advances Science</a>
    <a class="dropdown-item" href="../articles/b_course_part_1.html">B. Using R to Understand Bioinformatic Results</a>
    <a class="dropdown-item" href="../articles/c_course_part_2.html">C. R and Bioconductor for Genomic Analysis</a>
    <a class="dropdown-item" href="../articles/d_cxg.html">D. From CELLxGENE to R and Bioconductor</a>
    <a class="dropdown-item" href="../articles/e_sc_workflows.html">E. Seurat and Bioconductor Single-Cell Sequence Analysis Workflows</a>
    <a class="dropdown-item" href="../articles/f_osca_cell_annotation.html">F. Annotating cell types in Bioconductor</a>
    <a class="dropdown-item" href="../articles/g_osca_differential_expression.html">G. Differential Expression in Bioconductor</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mtmorgan/XM2023">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>F. Annotating cell types in Bioconductor</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mtmorgan/XM2023/blob/HEAD/vignettes/f_osca_cell_annotation.Rmd" class="external-link"><code>vignettes/f_osca_cell_annotation.Rmd</code></a></small>
      <div class="d-none name"><code>f_osca_cell_annotation.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="orchestrating-single-cell-analysis-with-bioconductor">Orchestrating Single-Cell Analysis with <em>Bioconductor</em><a class="anchor" aria-label="anchor" href="#orchestrating-single-cell-analysis-with-bioconductor"></a>
</h2>
<p><a href="https://bioconductor.org/books/OSCA" class="external-link">OSCA</a> – an amazing
resource!</p>
</div>
<div class="section level2">
<h2 id="annotating-cell-types">Annotating cell types<a class="anchor" aria-label="anchor" href="#annotating-cell-types"></a>
</h2>
<p>This script is derived from the <a href="https://bioconductor.org/books/OSCA.basic" class="external-link">OSCA Basic</a> book, <a href="https://bioconductor.org/books/release/OSCA.basic/cell-type-annotation.html" class="external-link">Chapter
7</a>: Cell type annotation. See the book for full details.</p>
<div class="section level3">
<h3 id="initial-analysis">Initial analysis<a class="anchor" aria-label="anchor" href="#initial-analysis"></a>
</h3>
<p>We start by retrieving some sample DropSeq data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DropletTestFiles</span><span class="op">)</span></span>
<span><span class="va">raw.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DropletTestFiles/man/listTestFiles.html" class="external-link">getTestFile</a></span><span class="op">(</span><span class="st">"tenx-2.1.0-pbmc4k/1.0.0/raw.tar.gz"</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?DropletTestFiles and browseVignettes('DropletTestFiles') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="va">out.path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"pbmc4k"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/untar.html" class="external-link">untar</a></span><span class="op">(</span><span class="va">raw.path</span>, exdir<span class="op">=</span><span class="va">out.path</span><span class="op">)</span></span></code></pre></div>
<p>Input the data using DropSeqUtils</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">DropletUtils</span><span class="op">)</span></span>
<span><span class="va">fname</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">out.path</span>, <span class="st">"raw_gene_bc_matrices/GRCh38"</span><span class="op">)</span></span>
<span><span class="va">sce.pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DropletUtils/man/read10xCounts.html" class="external-link">read10xCounts</a></span><span class="op">(</span><span class="va">fname</span>, col.names<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We lightly ‘annotate’ the data by making sure that feature names are
unique</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce.pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/uniquifyFeatureNames.html" class="external-link">uniquifyFeatureNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce.pbmc</span><span class="op">)</span><span class="op">$</span><span class="va">ID</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce.pbmc</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>As a quality control step, check for and remove empty droplets.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">e.out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DropletUtils/man/emptyDrops.html" class="external-link">emptyDrops</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce.pbmc</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce.pbmc</span> <span class="op">&lt;-</span> <span class="va">sce.pbmc</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">e.out</span><span class="op">$</span><span class="va">FDR</span> <span class="op">&lt;=</span> <span class="fl">0.001</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Normalize cell counts by clustering cells and computing scale factors
per cluster. Update the dataset to include log counts</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">quickCluster</a></span><span class="op">(</span><span class="va">sce.pbmc</span><span class="op">)</span></span>
<span><span class="va">sce.pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">sce.pbmc</span>, cluster<span class="op">=</span><span class="va">clusters</span><span class="op">)</span></span>
<span><span class="va">sce.pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce.pbmc</span><span class="op">)</span></span></code></pre></div>
<p>Variance modeling fits a statistical model to the count data; this is
used to identify ‘highly variable’ genes.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1001</span><span class="op">)</span></span>
<span><span class="va">dec.pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVarByPoisson.html" class="external-link">modelGeneVarByPoisson</a></span><span class="op">(</span><span class="va">sce.pbmc</span><span class="op">)</span></span>
<span><span class="va">top.pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec.pbmc</span>, prop<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p>Calculate reduced-dimensionality representations of the data.
<code><a href="https://rdrr.io/pkg/scran/man/denoisePCA.html" class="external-link">denoisePCA()</a></code> removes technical noise from log-normalized
counts. Technical noise, in contrast to ‘biological’ signal, is
variation between cells that is uncorrelated across genes. Removing
technical noise improves resolution in subsequent dimensionality
reduction steps. tSNE and UMAP are two common methods for reducing the
genes x samples data to two or a few dimensions for visual exploration.
Generally, UMAP is less sensitive to parameter choice and random number
seed.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">10000</span><span class="op">)</span></span>
<span><span class="va">sce.pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/denoisePCA.html" class="external-link">denoisePCA</a></span><span class="op">(</span><span class="va">sce.pbmc</span>, subset.row<span class="op">=</span><span class="va">top.pbmc</span>, technical<span class="op">=</span><span class="va">dec.pbmc</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100000</span><span class="op">)</span></span>
<span><span class="va">sce.pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce.pbmc</span>, dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000000</span><span class="op">)</span></span>
<span><span class="va">sce.pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce.pbmc</span>, dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></span></code></pre></div>
<p>Perform shared nearest-neighbor clustering
(<code><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph()</a></code>). These nearest neighbors are then used in
community detection algorithms (<code>cluster_walktrap()</code>) to find
similar cells. This represents a faster approach than more traditional
hierarchical clustering.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span><span class="va">sce.pbmc</span>, k<span class="op">=</span><span class="fl">10</span>, use.dimred <span class="op">=</span> <span class="st">'PCA'</span><span class="op">)</span></span>
<span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_walktrap.html" class="external-link">cluster_walktrap</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span><span class="op">$</span><span class="va">membership</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">sce.pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">clust</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="using-curated-reference-data">Using curated reference data<a class="anchor" aria-label="anchor" href="#using-curated-reference-data"></a>
</h3>
<p>One approach to cell annotation is to compare expression profiles to
a curated reference data set. We use the reference data derived from
Blueprint and ENCODE data; see <code>browseVignettes("celldex")</code>
for a description of this and other reference datasets.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu">celldex</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/celldex/man/BlueprintEncodeData.html" class="external-link">BlueprintEncodeData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?celldex and browseVignettes('celldex') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; see ?celldex and browseVignettes('celldex') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">ref</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 259 × 3</span></span></span>
<span><span class="co">#&gt;    label.main  label.fine  label.ont </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> Neutrophils Neutrophils CL:0000775</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Monocytes   Monocytes   CL:0000576</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Neutrophils Neutrophils CL:0000775</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> HSC         MEP         CL:0000050</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Neutrophils Neutrophils CL:0000775</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Monocytes   Monocytes   CL:0000576</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Neutrophils Neutrophils CL:0000775</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Monocytes   Monocytes   CL:0000576</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> Neutrophils Neutrophils CL:0000775</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Neutrophils Neutrophils CL:0000775</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 249 more rows</span></span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">ref</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">label.main</span>, sort <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 25 × 2</span></span></span>
<span><span class="co">#&gt;    label.main            n</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> HSC                  38</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> Endothelial cells    26</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> Macrophages          25</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> Neutrophils          23</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> Fibroblasts          20</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> Epithelial cells     18</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> Monocytes            16</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> Smooth muscle        16</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> CD4+ T-cells         14</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> Adipocytes            9</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 15 more rows</span></span></span></code></pre></div>
<p>Annotate our data using the ‘main’ label for cell type</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/SingleR" class="external-link">SingleR</a></span><span class="op">)</span></span>
<span><span class="co">## assign each cell in sce.pbmc to a type</span></span>
<span><span class="va">pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR</a></span><span class="op">(</span>test<span class="op">=</span><span class="va">sce.pbmc</span>, ref<span class="op">=</span><span class="va">ref</span>, labels<span class="op">=</span><span class="va">ref</span><span class="op">$</span><span class="va">label.main</span><span class="op">)</span></span>
<span><span class="va">pred</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"cell_id"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4,300 × 29</span></span></span>
<span><span class="co">#&gt;    cell_id            scores.Adipocytes scores.Astrocytes scores.B.cells</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> AAACCTGAGAAGGCCT-1             0.252            0.119           0.288</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> AAACCTGAGACAGACC-1             0.280            0.133           0.335</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> AAACCTGAGATAGTCA-1             0.267            0.149           0.298</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> AAACCTGAGGCATGGT-1             0.212            0.154           0.346</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> AAACCTGCAAGGTTCT-1             0.218            0.155           0.368</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> AAACCTGCAGGATTGG-1             0.244            0.139           0.382</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> AAACCTGCAGGCGATA-1             0.324            0.182           0.484</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> AAACCTGCATGAAGTA-1             0.313            0.167           0.461</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> AAACCTGGTAAATGAC-1             0.149            0.095<span style="text-decoration: underline;">5</span>          0.195</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> AAACCTGGTACATCCA-1             0.210            0.133           0.421</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 4,290 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 25 more variables: scores.CD4..T.cells &lt;dbl&gt;, scores.CD8..T.cells &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   scores.Chondrocytes &lt;dbl&gt;, scores.DC &lt;dbl&gt;, scores.Endothelial.cells &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   scores.Eosinophils &lt;dbl&gt;, scores.Epithelial.cells &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   scores.Erythrocytes &lt;dbl&gt;, scores.Fibroblasts &lt;dbl&gt;, scores.HSC &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   scores.Keratinocytes &lt;dbl&gt;, scores.Macrophages &lt;dbl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   scores.Melanocytes &lt;dbl&gt;, scores.Mesangial.cells &lt;dbl&gt;, …</span></span></span>
<span><span class="va">pred</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">labels</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 9 × 2</span></span></span>
<span><span class="co">#&gt;   labels           n</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> B-cells        578</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> CD4+ T-cells   815</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> CD8+ T-cells  <span style="text-decoration: underline;">1</span>358</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> DC               1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Eosinophils      1</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Erythrocytes     8</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">7</span> HSC             14</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">8</span> Monocytes     <span style="text-decoration: underline;">1</span>258</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">9</span> NK cells       267</span></span></code></pre></div>
<p>Finaly, visualize the cell annotation. Generally, B-cells and
Monocytes are well-annotated, e.g., cells labelled ‘B-cells’ have a
strong (yellow) signal only under the B-cell label; CD4+ and CD8+ cells
less well-anotated. It is interesting to explore biological reasons that
might contribute to this weaker classification.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/plotScoreHeatmap.html" class="external-link">plotScoreHeatmap</a></span><span class="op">(</span><span class="va">pred</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span>
<span></span>
<span><span class="co">#&gt; Warning: useNames = NA is deprecated. Instead, specify either useNames = TRUE</span></span>
<span><span class="co">#&gt; or useNames = TRUE.</span></span></code></pre></div>
<p><img src="f_osca_cell_annotation_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="other-strategies">Other strategies<a class="anchor" aria-label="anchor" href="#other-strategies"></a>
</h3>
<p>Other strategies ouline in <a href="https://bioconductor.org/books/release/OSCA.basic/cell-type-annotation.html" class="external-link">Chapter
7</a></p>
<ul>
<li>Custom references</li>
<li>Gene sets</li>
<li>Markers</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<p>This document was produced with the following <em>R</em>
software:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] SingleR_2.3.4               celldex_1.11.1             </span></span>
<span><span class="co">#&gt;  [3] scran_1.29.0                EnsDb.Hsapiens.v86_2.99.0  </span></span>
<span><span class="co">#&gt;  [5] ensembldb_2.25.0            AnnotationFilter_1.25.0    </span></span>
<span><span class="co">#&gt;  [7] GenomicFeatures_1.53.0      AnnotationDbi_1.63.1       </span></span>
<span><span class="co">#&gt;  [9] scater_1.29.0               ggplot2_3.4.2              </span></span>
<span><span class="co">#&gt; [11] scuttle_1.11.0              DropletUtils_1.21.0        </span></span>
<span><span class="co">#&gt; [13] SingleCellExperiment_1.23.0 SummarizedExperiment_1.31.1</span></span>
<span><span class="co">#&gt; [15] Biobase_2.61.0              GenomicRanges_1.53.1       </span></span>
<span><span class="co">#&gt; [17] GenomeInfoDb_1.37.1         IRanges_2.35.1             </span></span>
<span><span class="co">#&gt; [19] S4Vectors_0.39.1            BiocGenerics_0.47.0        </span></span>
<span><span class="co">#&gt; [21] MatrixGenerics_1.13.0       matrixStats_1.0.0          </span></span>
<span><span class="co">#&gt; [23] DropletTestFiles_1.11.0    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RcppAnnoy_0.0.20              later_1.3.1                  </span></span>
<span><span class="co">#&gt;   [3] BiocIO_1.11.0                 bitops_1.0-7                 </span></span>
<span><span class="co">#&gt;   [5] filelock_1.0.2                tibble_3.2.1                 </span></span>
<span><span class="co">#&gt;   [7] R.oo_1.25.0                   XML_3.99-0.14                </span></span>
<span><span class="co">#&gt;   [9] lifecycle_1.0.3               edgeR_3.43.4                 </span></span>
<span><span class="co">#&gt;  [11] rprojroot_2.0.3               lattice_0.21-8               </span></span>
<span><span class="co">#&gt;  [13] magrittr_2.0.3                limma_3.57.4                 </span></span>
<span><span class="co">#&gt;  [15] sass_0.4.6                    rmarkdown_2.22               </span></span>
<span><span class="co">#&gt;  [17] jquerylib_0.1.4               yaml_2.3.7                   </span></span>
<span><span class="co">#&gt;  [19] metapod_1.9.0                 httpuv_1.6.11                </span></span>
<span><span class="co">#&gt;  [21] RColorBrewer_1.1-3            DBI_1.1.3                    </span></span>
<span><span class="co">#&gt;  [23] zlibbioc_1.47.0               Rtsne_0.16                   </span></span>
<span><span class="co">#&gt;  [25] purrr_1.0.1                   R.utils_2.12.2               </span></span>
<span><span class="co">#&gt;  [27] RCurl_1.98-1.12               rappdirs_0.3.3               </span></span>
<span><span class="co">#&gt;  [29] GenomeInfoDbData_1.2.10       ggrepel_0.9.3                </span></span>
<span><span class="co">#&gt;  [31] irlba_2.3.5.1                 pheatmap_1.0.12              </span></span>
<span><span class="co">#&gt;  [33] dqrng_0.3.0                   pkgdown_2.0.7                </span></span>
<span><span class="co">#&gt;  [35] DelayedMatrixStats_1.23.0     codetools_0.2-19             </span></span>
<span><span class="co">#&gt;  [37] DelayedArray_0.27.5           xml2_1.3.4                   </span></span>
<span><span class="co">#&gt;  [39] tidyselect_1.2.0              farver_2.1.1                 </span></span>
<span><span class="co">#&gt;  [41] ScaledMatrix_1.9.1            viridis_0.6.3                </span></span>
<span><span class="co">#&gt;  [43] BiocFileCache_2.9.0           GenomicAlignments_1.37.0     </span></span>
<span><span class="co">#&gt;  [45] jsonlite_1.8.5                BiocNeighbors_1.19.0         </span></span>
<span><span class="co">#&gt;  [47] ellipsis_0.3.2                systemfonts_1.0.4            </span></span>
<span><span class="co">#&gt;  [49] tools_4.3.0                   progress_1.2.2               </span></span>
<span><span class="co">#&gt;  [51] ragg_1.2.5                    Rcpp_1.0.10                  </span></span>
<span><span class="co">#&gt;  [53] glue_1.6.2                    gridExtra_2.3                </span></span>
<span><span class="co">#&gt;  [55] SparseArray_1.1.10            xfun_0.39                    </span></span>
<span><span class="co">#&gt;  [57] dplyr_1.1.2                   HDF5Array_1.29.3             </span></span>
<span><span class="co">#&gt;  [59] withr_2.5.0                   BiocManager_1.30.21          </span></span>
<span><span class="co">#&gt;  [61] fastmap_1.1.1                 rhdf5filters_1.13.3          </span></span>
<span><span class="co">#&gt;  [63] bluster_1.11.1                fansi_1.0.4                  </span></span>
<span><span class="co">#&gt;  [65] digest_0.6.31                 rsvd_1.0.5                   </span></span>
<span><span class="co">#&gt;  [67] R6_2.5.1                      mime_0.12                    </span></span>
<span><span class="co">#&gt;  [69] textshaping_0.3.6             colorspace_2.1-0             </span></span>
<span><span class="co">#&gt;  [71] biomaRt_2.57.1                RSQLite_2.3.1                </span></span>
<span><span class="co">#&gt;  [73] R.methodsS3_1.8.2             utf8_1.2.3                   </span></span>
<span><span class="co">#&gt;  [75] generics_0.1.3                rtracklayer_1.61.0           </span></span>
<span><span class="co">#&gt;  [77] prettyunits_1.1.1             httr_1.4.6                   </span></span>
<span><span class="co">#&gt;  [79] S4Arrays_1.1.4                uwot_0.1.14                  </span></span>
<span><span class="co">#&gt;  [81] pkgconfig_2.0.3               gtable_0.3.3                 </span></span>
<span><span class="co">#&gt;  [83] blob_1.2.4                    XVector_0.41.1               </span></span>
<span><span class="co">#&gt;  [85] htmltools_0.5.5               ProtGenerics_1.33.0          </span></span>
<span><span class="co">#&gt;  [87] scales_1.2.1                  png_0.1-8                    </span></span>
<span><span class="co">#&gt;  [89] knitr_1.43                    rjson_0.2.21                 </span></span>
<span><span class="co">#&gt;  [91] curl_5.0.1                    cachem_1.0.8                 </span></span>
<span><span class="co">#&gt;  [93] rhdf5_2.45.0                  stringr_1.5.0                </span></span>
<span><span class="co">#&gt;  [95] BiocVersion_3.18.0            parallel_4.3.0               </span></span>
<span><span class="co">#&gt;  [97] vipor_0.4.5                   restfulr_0.0.15              </span></span>
<span><span class="co">#&gt;  [99] desc_1.4.2                    pillar_1.9.0                 </span></span>
<span><span class="co">#&gt; [101] grid_4.3.0                    vctrs_0.6.3                  </span></span>
<span><span class="co">#&gt; [103] promises_1.2.0.1              BiocSingular_1.17.0          </span></span>
<span><span class="co">#&gt; [105] dbplyr_2.3.2                  beachmat_2.17.8              </span></span>
<span><span class="co">#&gt; [107] xtable_1.8-4                  cluster_2.1.4                </span></span>
<span><span class="co">#&gt; [109] beeswarm_0.4.0                evaluate_0.21                </span></span>
<span><span class="co">#&gt; [111] cli_3.6.1                     locfit_1.5-9.8               </span></span>
<span><span class="co">#&gt; [113] compiler_4.3.0                Rsamtools_2.17.0             </span></span>
<span><span class="co">#&gt; [115] rlang_1.1.1                   crayon_1.5.2                 </span></span>
<span><span class="co">#&gt; [117] fs_1.6.2                      ggbeeswarm_0.7.2             </span></span>
<span><span class="co">#&gt; [119] stringi_1.7.12                viridisLite_0.4.2            </span></span>
<span><span class="co">#&gt; [121] BiocParallel_1.35.2           munsell_0.5.0                </span></span>
<span><span class="co">#&gt; [123] Biostrings_2.69.1             lazyeval_0.2.2               </span></span>
<span><span class="co">#&gt; [125] Matrix_1.5-4.1                ExperimentHub_2.7.1          </span></span>
<span><span class="co">#&gt; [127] hms_1.1.3                     sparseMatrixStats_1.13.0     </span></span>
<span><span class="co">#&gt; [129] bit64_4.0.5                   Rhdf5lib_1.23.0              </span></span>
<span><span class="co">#&gt; [131] KEGGREST_1.41.0               statmod_1.5.0                </span></span>
<span><span class="co">#&gt; [133] shiny_1.7.4                   highr_0.10                   </span></span>
<span><span class="co">#&gt; [135] interactiveDisplayBase_1.39.0 AnnotationHub_3.9.1          </span></span>
<span><span class="co">#&gt; [137] igraph_1.4.3                  memoise_2.0.1                </span></span>
<span><span class="co">#&gt; [139] bslib_0.5.0                   bit_4.0.5</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Martin Morgan, NIH NHGRI U24HG004059.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
