<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="XM2023">
<title>G. Differential Expression in Bioconductor • XM2023</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="G. Differential Expression in Bioconductor">
<meta property="og:description" content="XM2023">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">XM2023</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9012</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/a_talk.html">A. How Bioconductor Advances Science</a>
    <a class="dropdown-item" href="../articles/b_course_part_1.html">B. Using R to Understand Bioinformatic Results</a>
    <a class="dropdown-item" href="../articles/c_course_part_2.html">C. R and Bioconductor for Genomic Analysis</a>
    <a class="dropdown-item" href="../articles/d_cxg.html">D. From CELLxGENE to R and Bioconductor</a>
    <a class="dropdown-item" href="../articles/e_sc_workflows.html">E. Seurat and Bioconductor Single-Cell Sequence Analysis Workflows</a>
    <a class="dropdown-item" href="../articles/f_osca_cell_annotation.html">F. Annotating cell types in Bioconductor</a>
    <a class="dropdown-item" href="../articles/g_osca_differential_expression.html">G. Differential Expression in Bioconductor</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/mtmorgan/XM2023">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>G. Differential Expression in Bioconductor</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/mtmorgan/XM2023/blob/HEAD/vignettes/g_osca_differential_expression.Rmd" class="external-link"><code>vignettes/g_osca_differential_expression.Rmd</code></a></small>
      <div class="d-none name"><code>g_osca_differential_expression.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="orchestrating-single-cell-analysis-with-bioconductor">Orchestrating Single-Cell Analysis with <em>Bioconductor</em><a class="anchor" aria-label="anchor" href="#orchestrating-single-cell-analysis-with-bioconductor"></a>
</h2>
<p><a href="https://bioconductor.org/books/OSCA" class="external-link">OSCA</a> – an amazing
resource!</p>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p>This script is derived from the <a href="https://bioconductor.org/books/OSCA.multisample/" class="external-link">OSCA
Multi-Sample</a> book, <a href="https://bioconductor.org/books/release/OSCA.multisample/multi-sample-comparisons.html" class="external-link">Chapter
4</a>: DE analyses between conditions. See the book for full
details.</p>
<div class="section level3">
<h3 id="initial-analysis">Initial analysis<a class="anchor" aria-label="anchor" href="#initial-analysis"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## load data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/MouseGastrulationData" class="external-link">MouseGastrulationData</a></span><span class="op">)</span></span>
<span><span class="va">sce.chimera</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MouseGastrulationData/man/WTChimeraData.html" class="external-link">WTChimeraData</a></span><span class="op">(</span>samples<span class="op">=</span><span class="fl">5</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">sce.chimera</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 29453 20935 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000095742 tomato-td</span></span>
<span><span class="co">#&gt; rowData names(2): ENSEMBL SYMBOL</span></span>
<span><span class="co">#&gt; colnames(20935): cell_9769 cell_9770 ... cell_30702 cell_30703</span></span>
<span><span class="co">#&gt; colData names(11): cell barcode ... doub.density sizeFactor</span></span>
<span><span class="co">#&gt; reducedDimNames(2): pca.corrected.E7.5 pca.corrected.E8.5</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span>
<span></span>
<span><span class="co">## feature annotation</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce.chimera</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/uniquifyFeatureNames.html" class="external-link">uniquifyFeatureNames</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce.chimera</span><span class="op">)</span><span class="op">$</span><span class="va">ENSEMBL</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce.chimera</span><span class="op">)</span><span class="op">$</span><span class="va">SYMBOL</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## qc</span></span>
<span><span class="va">drop</span> <span class="op">&lt;-</span> <span class="va">sce.chimera</span><span class="op">$</span><span class="va">celltype.mapped</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"stripped"</span>, <span class="st">"Doublet"</span><span class="op">)</span></span>
<span><span class="va">sce.chimera</span> <span class="op">&lt;-</span> <span class="va">sce.chimera</span><span class="op">[</span>,<span class="op">!</span><span class="va">drop</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## normalization</span></span>
<span><span class="va">sce.chimera</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce.chimera</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## variance modeling</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="va">dec.chimera</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce.chimera</span>, block<span class="op">=</span><span class="va">sce.chimera</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span><span class="va">chosen.hvgs</span> <span class="op">&lt;-</span> <span class="va">dec.chimera</span><span class="op">$</span><span class="va">bio</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span></span>
<span><span class="co">## merge batches</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">01001001</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/correctExperiments.html" class="external-link">correctExperiments</a></span><span class="op">(</span></span>
<span>        <span class="va">sce.chimera</span>,</span>
<span>        batch<span class="op">=</span><span class="va">sce.chimera</span><span class="op">$</span><span class="va">sample</span>,</span>
<span>        subset.row<span class="op">=</span><span class="va">chosen.hvgs</span>,</span>
<span>        PARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/BatchelorParam.html" class="external-link">FastMnnParam</a></span><span class="op">(</span></span>
<span>            merge.order<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">5</span><span class="op">)</span>, <span class="co"># WT (3 replicates)</span></span>
<span>                <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">6</span><span class="op">)</span>  <span class="co"># td-Tomato (3 replicates)</span></span>
<span>           <span class="op">)</span></span>
<span>       <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## cluster</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span><span class="va">merged</span>, use.dimred<span class="op">=</span><span class="st">"corrected"</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_louvain.html" class="external-link">cluster_louvain</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">merged</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">$</span><span class="va">membership</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## dimensionality reduction</span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">merged</span>, dimred<span class="op">=</span><span class="st">"corrected"</span>, external_neighbors<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">merged</span>, dimred<span class="op">=</span><span class="st">"corrected"</span>, external_neighbors<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="differential-expression-1">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression-1"></a>
</h3>
<p>Initial observations – labels are similar, whether a ‘td-Tomato’
positive cell or not.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">merged</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">label</span>, <span class="va">tomato</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_wider.html" class="external-link">pivot_wider</a></span><span class="op">(</span>names_from <span class="op">=</span> <span class="st">"tomato"</span>, values_from <span class="op">=</span> <span class="st">"n"</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 26 × 3</span></span></span>
<span><span class="co">#&gt;    label `FALSE` `TRUE`</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 1         129    264</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 2         363    512</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 3         417    632</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 4        <span style="text-decoration: underline;">1</span>255    973</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 5        <span style="text-decoration: underline;">1</span>444    753</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 6         868   <span style="text-decoration: underline;">1</span>092</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 7         542    398</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 8         476    405</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 9         269    333</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 10        608    604</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 16 more rows</span></span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">merged</span>, colour_by<span class="op">=</span><span class="st">"tomato"</span>, text_by<span class="op">=</span><span class="st">"label"</span>, point_size <span class="op">=</span> <span class="fl">.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="g_osca_differential_expression_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>Create pseudo-bulk samples</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using 'label' and 'sample' as our two factors; each column of the output</span></span>
<span><span class="co"># corresponds to one unique combination of these two factors.</span></span>
<span><span class="va">summed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/aggregateAcrossCells.html" class="external-link">aggregateAcrossCells</a></span><span class="op">(</span></span>
<span>    <span class="va">merged</span>, </span>
<span>    id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">merged</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"celltype.mapped"</span>, <span class="st">"sample"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="va">summed</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 14699 186 </span></span>
<span><span class="co">#&gt; metadata(2): merge.info pca.info</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(14699): Xkr4 Rp1 ... Vmn2r122 CAAA01147332.1</span></span>
<span><span class="co">#&gt; rowData names(3): rotation ENSEMBL SYMBOL</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(16): batch cell ... sample ncells</span></span>
<span><span class="co">#&gt; reducedDimNames(5): corrected pca.corrected.E7.5 pca.corrected.E8.5</span></span>
<span><span class="co">#&gt;   TSNE UMAP</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="conclusions">Conclusions<a class="anchor" aria-label="anchor" href="#conclusions"></a>
</h2>
<p>Check out the following for a careful and comprehensive evaluation of
single-cell differential expression methods:</p>
<p>Soneson, C., Robinson, M. Bias, robustness and scalability in
single-cell differential expression analysis. Nat Methods 15, 255–261
(2018). <a href="https://doi.org/10.1038/nmeth.4612" class="external-link uri">https://doi.org/10.1038/nmeth.4612</a></p>
<div class="section level3">
<h3 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h3>
<p>This document was produced with the following <em>R</em>
software:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: Etc/UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] batchelor_1.17.0             scran_1.29.0                </span></span>
<span><span class="co">#&gt;  [3] scater_1.29.0                ggplot2_3.4.2               </span></span>
<span><span class="co">#&gt;  [5] scuttle_1.11.0               MouseGastrulationData_1.15.0</span></span>
<span><span class="co">#&gt;  [7] SpatialExperiment_1.11.0     SingleCellExperiment_1.23.0 </span></span>
<span><span class="co">#&gt;  [9] SummarizedExperiment_1.31.1  Biobase_2.61.0              </span></span>
<span><span class="co">#&gt; [11] GenomicRanges_1.53.1         GenomeInfoDb_1.37.1         </span></span>
<span><span class="co">#&gt; [13] IRanges_2.35.1               S4Vectors_0.39.1            </span></span>
<span><span class="co">#&gt; [15] BiocGenerics_0.47.0          MatrixGenerics_1.13.0       </span></span>
<span><span class="co">#&gt; [17] matrixStats_1.0.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] jsonlite_1.8.5                magrittr_2.0.3               </span></span>
<span><span class="co">#&gt;   [3] ggbeeswarm_0.7.2              magick_2.7.4                 </span></span>
<span><span class="co">#&gt;   [5] farver_2.1.1                  rmarkdown_2.22               </span></span>
<span><span class="co">#&gt;   [7] fs_1.6.2                      zlibbioc_1.47.0              </span></span>
<span><span class="co">#&gt;   [9] ragg_1.2.5                    vctrs_0.6.3                  </span></span>
<span><span class="co">#&gt;  [11] memoise_2.0.1                 DelayedMatrixStats_1.23.0    </span></span>
<span><span class="co">#&gt;  [13] RCurl_1.98-1.12               htmltools_0.5.5              </span></span>
<span><span class="co">#&gt;  [15] S4Arrays_1.1.4                AnnotationHub_3.9.1          </span></span>
<span><span class="co">#&gt;  [17] curl_5.0.1                    BiocNeighbors_1.19.0         </span></span>
<span><span class="co">#&gt;  [19] Rhdf5lib_1.23.0               SparseArray_1.1.10           </span></span>
<span><span class="co">#&gt;  [21] rhdf5_2.45.0                  sass_0.4.6                   </span></span>
<span><span class="co">#&gt;  [23] bslib_0.5.0                   desc_1.4.2                   </span></span>
<span><span class="co">#&gt;  [25] cachem_1.0.8                  ResidualMatrix_1.11.0        </span></span>
<span><span class="co">#&gt;  [27] igraph_1.4.3                  mime_0.12                    </span></span>
<span><span class="co">#&gt;  [29] lifecycle_1.0.3               pkgconfig_2.0.3              </span></span>
<span><span class="co">#&gt;  [31] rsvd_1.0.5                    Matrix_1.5-4.1               </span></span>
<span><span class="co">#&gt;  [33] R6_2.5.1                      fastmap_1.1.1                </span></span>
<span><span class="co">#&gt;  [35] GenomeInfoDbData_1.2.10       shiny_1.7.4                  </span></span>
<span><span class="co">#&gt;  [37] digest_0.6.31                 colorspace_2.1-0             </span></span>
<span><span class="co">#&gt;  [39] AnnotationDbi_1.63.1          rprojroot_2.0.3              </span></span>
<span><span class="co">#&gt;  [41] dqrng_0.3.0                   irlba_2.3.5.1                </span></span>
<span><span class="co">#&gt;  [43] ExperimentHub_2.7.1           textshaping_0.3.6            </span></span>
<span><span class="co">#&gt;  [45] RSQLite_2.3.1                 beachmat_2.17.8              </span></span>
<span><span class="co">#&gt;  [47] labeling_0.4.2                filelock_1.0.2               </span></span>
<span><span class="co">#&gt;  [49] fansi_1.0.4                   httr_1.4.6                   </span></span>
<span><span class="co">#&gt;  [51] compiler_4.3.0                bit64_4.0.5                  </span></span>
<span><span class="co">#&gt;  [53] withr_2.5.0                   BiocParallel_1.35.2          </span></span>
<span><span class="co">#&gt;  [55] viridis_0.6.3                 DBI_1.1.3                    </span></span>
<span><span class="co">#&gt;  [57] highr_0.10                    HDF5Array_1.29.3             </span></span>
<span><span class="co">#&gt;  [59] R.utils_2.12.2                rappdirs_0.3.3               </span></span>
<span><span class="co">#&gt;  [61] DelayedArray_0.27.5           bluster_1.11.1               </span></span>
<span><span class="co">#&gt;  [63] rjson_0.2.21                  tools_4.3.0                  </span></span>
<span><span class="co">#&gt;  [65] vipor_0.4.5                   beeswarm_0.4.0               </span></span>
<span><span class="co">#&gt;  [67] interactiveDisplayBase_1.39.0 httpuv_1.6.11                </span></span>
<span><span class="co">#&gt;  [69] R.oo_1.25.0                   glue_1.6.2                   </span></span>
<span><span class="co">#&gt;  [71] rhdf5filters_1.13.3           promises_1.2.0.1             </span></span>
<span><span class="co">#&gt;  [73] grid_4.3.0                    Rtsne_0.16                   </span></span>
<span><span class="co">#&gt;  [75] cluster_2.1.4                 generics_0.1.3               </span></span>
<span><span class="co">#&gt;  [77] gtable_0.3.3                  R.methodsS3_1.8.2            </span></span>
<span><span class="co">#&gt;  [79] tidyr_1.3.0                   metapod_1.9.0                </span></span>
<span><span class="co">#&gt;  [81] BiocSingular_1.17.0           ScaledMatrix_1.9.1           </span></span>
<span><span class="co">#&gt;  [83] utf8_1.2.3                    XVector_0.41.1               </span></span>
<span><span class="co">#&gt;  [85] ggrepel_0.9.3                 BiocVersion_3.18.0           </span></span>
<span><span class="co">#&gt;  [87] pillar_1.9.0                  stringr_1.5.0                </span></span>
<span><span class="co">#&gt;  [89] limma_3.57.4                  BumpyMatrix_1.9.0            </span></span>
<span><span class="co">#&gt;  [91] later_1.3.1                   dplyr_1.1.2                  </span></span>
<span><span class="co">#&gt;  [93] BiocFileCache_2.9.0           lattice_0.21-8               </span></span>
<span><span class="co">#&gt;  [95] bit_4.0.5                     tidyselect_1.2.0             </span></span>
<span><span class="co">#&gt;  [97] locfit_1.5-9.8                Biostrings_2.69.1            </span></span>
<span><span class="co">#&gt;  [99] knitr_1.43                    gridExtra_2.3                </span></span>
<span><span class="co">#&gt; [101] edgeR_3.43.4                  xfun_0.39                    </span></span>
<span><span class="co">#&gt; [103] statmod_1.5.0                 DropletUtils_1.21.0          </span></span>
<span><span class="co">#&gt; [105] stringi_1.7.12                yaml_2.3.7                   </span></span>
<span><span class="co">#&gt; [107] evaluate_0.21                 codetools_0.2-19             </span></span>
<span><span class="co">#&gt; [109] tibble_3.2.1                  BiocManager_1.30.21          </span></span>
<span><span class="co">#&gt; [111] cli_3.6.1                     uwot_0.1.14                  </span></span>
<span><span class="co">#&gt; [113] xtable_1.8-4                  systemfonts_1.0.4            </span></span>
<span><span class="co">#&gt; [115] munsell_0.5.0                 jquerylib_0.1.4              </span></span>
<span><span class="co">#&gt; [117] Rcpp_1.0.10                   dbplyr_2.3.2                 </span></span>
<span><span class="co">#&gt; [119] png_0.1-8                     parallel_4.3.0               </span></span>
<span><span class="co">#&gt; [121] ellipsis_0.3.2                pkgdown_2.0.7                </span></span>
<span><span class="co">#&gt; [123] blob_1.2.4                    sparseMatrixStats_1.13.0     </span></span>
<span><span class="co">#&gt; [125] bitops_1.0-7                  viridisLite_0.4.2            </span></span>
<span><span class="co">#&gt; [127] scales_1.2.1                  purrr_1.0.1                  </span></span>
<span><span class="co">#&gt; [129] crayon_1.5.2                  rlang_1.1.1                  </span></span>
<span><span class="co">#&gt; [131] cowplot_1.1.1                 KEGGREST_1.41.0</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Martin Morgan, NIH NHGRI U24HG004059.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
